{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "# Cancer Genome Simulation Overview\n",
    "This notebook provides a pipeline to run insilicoSV sequentially to simulate cancer genomes.\n",
    "This notebook takes as input a YAML file containing:\n",
    "  - The paths to several insilicoSV YAML config files.\n",
    "  - A list of the tumor clones and the sequence of config files to run to obtain the clone genome.\n",
    "  - The tumor purity of each clone for the simulation of the reads.\n",
    "Refer to the provided clones.yaml config for an example of the expected syntax. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": "## Generate the Clone Genomes"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:54.158335Z",
     "start_time": "2025-08-12T01:33:54.152545Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import sys\n",
    "\n",
    "import yaml\n",
    "from pysam import VariantFile\n",
    "from IPython.display import Image\n",
    "from collections import defaultdict\n",
    "import os\n",
    "import subprocess\n",
    "import shutil\n",
    "from math import ceil"
   ],
   "outputs": [],
   "execution_count": 67
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:54.174510Z",
     "start_time": "2025-08-12T01:33:54.171829Z"
    }
   },
   "cell_type": "code",
   "source": "number_threads = '10'",
   "outputs": [],
   "execution_count": 68
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:58.681360Z",
     "start_time": "2025-08-12T01:33:54.186364Z"
    }
   },
   "source": [
    "%%sh\n",
    "rm -r ./clones/\n",
    "mkdir -p clones\n",
    "\n",
    "# download the chr21 reference\n",
    "wget -O clones/chr21.fa.gz https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr21.fa.gz\n",
    "gunzip -f clones/chr21.fa.gz"
   ],
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "--2025-08-11 21:33:55--  https://hgdownload.soe.ucsc.edu/goldenPath/hg38/chromosomes/chr21.fa.gz\n",
      "Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163\n",
      "Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 12709705 (12M) [application/x-gzip]\n",
      "Saving to: ‘clones/chr21.fa.gz’\n",
      "\n",
      "     0K .......... .......... .......... .......... ..........  0%  325K 38s\n",
      "    50K .......... .......... .......... .......... ..........  0%  653K 28s\n",
      "   100K .......... .......... .......... .......... ..........  1% 8.72M 19s\n",
      "   150K .......... .......... .......... .......... ..........  1% 33.3M 14s\n",
      "   200K .......... .......... .......... .......... ..........  2%  671K 15s\n",
      "   250K .......... .......... .......... .......... ..........  2% 14.5M 13s\n",
      "   300K .......... .......... .......... .......... ..........  2%  152M 11s\n",
      "   350K .......... .......... .......... .......... ..........  3% 19.3M 10s\n",
      "   400K .......... .......... .......... .......... ..........  3%  688K 10s\n",
      "   450K .......... .......... .......... .......... ..........  4% 6.28M 9s\n",
      "   500K .......... .......... .......... .......... ..........  4% 2.47M 9s\n",
      "   550K .......... .......... .......... .......... ..........  4% 4.04M 8s\n",
      "   600K .......... .......... .......... .......... ..........  5% 8.21M 8s\n",
      "   650K .......... .......... .......... .......... ..........  5% 6.63M 7s\n",
      "   700K .......... .......... .......... .......... ..........  6% 23.5M 7s\n",
      "   750K .......... .......... .......... .......... ..........  6% 13.8M 7s\n",
      "   800K .......... .......... .......... .......... ..........  6% 12.9M 6s\n",
      "   850K .......... .......... .......... .......... ..........  7% 2.66M 6s\n",
      "   900K .......... .......... .......... .......... ..........  7% 4.12M 6s\n",
      "   950K .......... .......... .......... .......... ..........  8% 5.04M 6s\n",
      "  1000K .......... .......... .......... .......... ..........  8% 1.06M 6s\n",
      "  1050K .......... .......... .......... .......... ..........  8% 11.6M 6s\n",
      "  1100K .......... .......... .......... .......... ..........  9% 23.3M 5s\n",
      "  1150K .......... .......... .......... .......... ..........  9% 3.26M 5s\n",
      "  1200K .......... .......... .......... .......... .......... 10%  379M 5s\n",
      "  1250K .......... .......... .......... .......... .......... 10% 2.83M 5s\n",
      "  1300K .......... .......... .......... .......... .......... 10% 11.9M 5s\n",
      "  1350K .......... .......... .......... .......... .......... 11% 4.48M 5s\n",
      "  1400K .......... .......... .......... .......... .......... 11% 2.59M 5s\n",
      "  1450K .......... .......... .......... .......... .......... 12% 9.91M 4s\n",
      "  1500K .......... .......... .......... .......... .......... 12% 7.79M 4s\n",
      "  1550K .......... .......... .......... .......... .......... 12% 12.8M 4s\n",
      "  1600K .......... .......... .......... .......... .......... 13% 18.8M 4s\n",
      "  1650K .......... .......... .......... .......... .......... 13% 19.4M 4s\n",
      "  1700K .......... .......... .......... .......... .......... 14% 14.3M 4s\n",
      "  1750K .......... .......... .......... .......... .......... 14% 54.4M 4s\n",
      "  1800K .......... .......... .......... .......... .......... 14% 14.5M 4s\n",
      "  1850K .......... .......... .......... .......... .......... 15% 22.5M 4s\n",
      "  1900K .......... .......... .......... .......... .......... 15% 25.5M 3s\n",
      "  1950K .......... .......... .......... .......... .......... 16% 10.8M 3s\n",
      "  2000K .......... .......... .......... .......... .......... 16%  104M 3s\n",
      "  2050K .......... .......... .......... .......... .......... 16% 18.9M 3s\n",
      "  2100K .......... .......... .......... .......... .......... 17% 13.1M 3s\n",
      "  2150K .......... .......... .......... .......... .......... 17% 18.8M 3s\n",
      "  2200K .......... .......... .......... .......... .......... 18% 31.3M 3s\n",
      "  2250K .......... .......... .......... .......... .......... 18% 9.85M 3s\n",
      "  2300K .......... .......... .......... .......... .......... 18% 13.0M 3s\n",
      "  2350K .......... .......... .......... .......... .......... 19% 9.72M 3s\n",
      "  2400K .......... .......... .......... .......... .......... 19% 15.8M 3s\n",
      "  2450K .......... .......... .......... .......... .......... 20% 29.4M 3s\n",
      "  2500K .......... .......... .......... .......... .......... 20% 23.7M 3s\n",
      "  2550K .......... .......... .......... .......... .......... 20% 24.8M 3s\n",
      "  2600K .......... .......... .......... .......... .......... 21% 16.2M 3s\n",
      "  2650K .......... .......... .......... .......... .......... 21% 14.4M 2s\n",
      "  2700K .......... .......... .......... .......... .......... 22%  110M 2s\n",
      "  2750K .......... .......... .......... .......... .......... 22% 28.3M 2s\n",
      "  2800K .......... .......... .......... .......... .......... 22% 11.6M 2s\n",
      "  2850K .......... .......... .......... .......... .......... 23% 1.43M 2s\n",
      "  2900K .......... .......... .......... .......... .......... 23% 19.7M 2s\n",
      "  2950K .......... .......... .......... .......... .......... 24% 11.9M 2s\n",
      "  3000K .......... .......... .......... .......... .......... 24% 12.7M 2s\n",
      "  3050K .......... .......... .......... .......... .......... 24% 43.0M 2s\n",
      "  3100K .......... .......... .......... .......... .......... 25%  188M 2s\n",
      "  3150K .......... .......... .......... .......... .......... 25% 26.9M 2s\n",
      "  3200K .......... .......... .......... .......... .......... 26% 7.00M 2s\n",
      "  3250K .......... .......... .......... .......... .......... 26% 52.7M 2s\n",
      "  3300K .......... .......... .......... .......... .......... 26% 16.9M 2s\n",
      "  3350K .......... .......... .......... .......... .......... 27% 12.5M 2s\n",
      "  3400K .......... .......... .......... .......... .......... 27% 34.2M 2s\n",
      "  3450K .......... .......... .......... .......... .......... 28% 19.5M 2s\n",
      "  3500K .......... .......... .......... .......... .......... 28% 16.8M 2s\n",
      "  3550K .......... .......... .......... .......... .......... 29% 42.2M 2s\n",
      "  3600K .......... .......... .......... .......... .......... 29% 22.9M 2s\n",
      "  3650K .......... .......... .......... .......... .......... 29% 18.0M 2s\n",
      "  3700K .......... .......... .......... .......... .......... 30% 21.2M 2s\n",
      "  3750K .......... .......... .......... .......... .......... 30% 15.7M 2s\n",
      "  3800K .......... .......... .......... .......... .......... 31% 20.5M 2s\n",
      "  3850K .......... .......... .......... .......... .......... 31% 51.8M 2s\n",
      "  3900K .......... .......... .......... .......... .......... 31% 13.0M 2s\n",
      "  3950K .......... .......... .......... .......... .......... 32% 18.8M 2s\n",
      "  4000K .......... .......... .......... .......... .......... 32% 23.6M 2s\n",
      "  4050K .......... .......... .......... .......... .......... 33% 24.0M 2s\n",
      "  4100K .......... .......... .......... .......... .......... 33% 16.0M 2s\n",
      "  4150K .......... .......... .......... .......... .......... 33% 29.0M 2s\n",
      "  4200K .......... .......... .......... .......... .......... 34% 18.1M 2s\n",
      "  4250K .......... .......... .......... .......... .......... 34% 31.1M 2s\n",
      "  4300K .......... .......... .......... .......... .......... 35% 24.5M 1s\n",
      "  4350K .......... .......... .......... .......... .......... 35% 12.0M 1s\n",
      "  4400K .......... .......... .......... .......... .......... 35% 22.7M 1s\n",
      "  4450K .......... .......... .......... .......... .......... 36% 9.93M 1s\n",
      "  4500K .......... .......... .......... .......... .......... 36% 10.7M 1s\n",
      "  4550K .......... .......... .......... .......... .......... 37% 38.7M 1s\n",
      "  4600K .......... .......... .......... .......... .......... 37% 11.1M 1s\n",
      "  4650K .......... .......... .......... .......... .......... 37% 42.6M 1s\n",
      "  4700K .......... .......... .......... .......... .......... 38% 15.2M 1s\n",
      "  4750K .......... .......... .......... .......... .......... 38% 3.68M 1s\n",
      "  4800K .......... .......... .......... .......... .......... 39% 58.1M 1s\n",
      "  4850K .......... .......... .......... .......... .......... 39% 14.5M 1s\n",
      "  4900K .......... .......... .......... .......... .......... 39% 22.2M 1s\n",
      "  4950K .......... .......... .......... .......... .......... 40% 21.9M 1s\n",
      "  5000K .......... .......... .......... .......... .......... 40% 18.0M 1s\n",
      "  5050K .......... .......... .......... .......... .......... 41% 36.6M 1s\n",
      "  5100K .......... .......... .......... .......... .......... 41% 20.1M 1s\n",
      "  5150K .......... .......... .......... .......... .......... 41% 23.5M 1s\n",
      "  5200K .......... .......... .......... .......... .......... 42% 32.6M 1s\n",
      "  5250K .......... .......... .......... .......... .......... 42% 14.4M 1s\n",
      "  5300K .......... .......... .......... .......... .......... 43% 25.5M 1s\n",
      "  5350K .......... .......... .......... .......... .......... 43% 23.0M 1s\n",
      "  5400K .......... .......... .......... .......... .......... 43% 12.3M 1s\n",
      "  5450K .......... .......... .......... .......... .......... 44% 22.3M 1s\n",
      "  5500K .......... .......... .......... .......... .......... 44% 24.0M 1s\n",
      "  5550K .......... .......... .......... .......... .......... 45% 4.53M 1s\n",
      "  5600K .......... .......... .......... .......... .......... 45% 6.52M 1s\n",
      "  5650K .......... .......... .......... .......... .......... 45% 24.8M 1s\n",
      "  5700K .......... .......... .......... .......... .......... 46%  208M 1s\n",
      "  5750K .......... .......... .......... .......... .......... 46% 98.8M 1s\n",
      "  5800K .......... .......... .......... .......... .......... 47%  112M 1s\n",
      "  5850K .......... .......... .......... .......... .......... 47% 10.8M 1s\n",
      "  5900K .......... .......... .......... .......... .......... 47%  195M 1s\n",
      "  5950K .......... .......... .......... .......... .......... 48% 25.5M 1s\n",
      "  6000K .......... .......... .......... .......... .......... 48% 51.8M 1s\n",
      "  6050K .......... .......... .......... .......... .......... 49% 14.3M 1s\n",
      "  6100K .......... .......... .......... .......... .......... 49% 18.6M 1s\n",
      "  6150K .......... .......... .......... .......... .......... 49% 32.9M 1s\n",
      "  6200K .......... .......... .......... .......... .......... 50% 24.5M 1s\n",
      "  6250K .......... .......... .......... .......... .......... 50% 17.5M 1s\n",
      "  6300K .......... .......... .......... .......... .......... 51% 34.5M 1s\n",
      "  6350K .......... .......... .......... .......... .......... 51% 12.1M 1s\n",
      "  6400K .......... .......... .......... .......... .......... 51% 12.1M 1s\n",
      "  6450K .......... .......... .......... .......... .......... 52%  237M 1s\n",
      "  6500K .......... .......... .......... .......... .......... 52% 22.6M 1s\n",
      "  6550K .......... .......... .......... .......... .......... 53% 15.9M 1s\n",
      "  6600K .......... .......... .......... .......... .......... 53% 27.5M 1s\n",
      "  6650K .......... .......... .......... .......... .......... 53% 25.9M 1s\n",
      "  6700K .......... .......... .......... .......... .......... 54% 11.3M 1s\n",
      "  6750K .......... .......... .......... .......... .......... 54% 18.9M 1s\n",
      "  6800K .......... .......... .......... .......... .......... 55% 17.9M 1s\n",
      "  6850K .......... .......... .......... .......... .......... 55% 20.9M 1s\n",
      "  6900K .......... .......... .......... .......... .......... 55% 59.3M 1s\n",
      "  6950K .......... .......... .......... .......... .......... 56% 15.4M 1s\n",
      "  7000K .......... .......... .......... .......... .......... 56% 22.1M 1s\n",
      "  7050K .......... .......... .......... .......... .......... 57% 28.9M 1s\n",
      "  7100K .......... .......... .......... .......... .......... 57% 4.11M 1s\n",
      "  7150K .......... .......... .......... .......... .......... 58% 13.5M 1s\n",
      "  7200K .......... .......... .......... .......... .......... 58% 94.8M 1s\n",
      "  7250K .......... .......... .......... .......... .......... 58%  122M 1s\n",
      "  7300K .......... .......... .......... .......... .......... 59% 36.9M 1s\n",
      "  7350K .......... .......... .......... .......... .......... 59%  397M 1s\n",
      "  7400K .......... .......... .......... .......... .......... 60%  519M 1s\n",
      "  7450K .......... .......... .......... .......... .......... 60% 3.74M 1s\n",
      "  7500K .......... .......... .......... .......... .......... 60% 2.54M 1s\n",
      "  7550K .......... .......... .......... .......... .......... 61% 3.13M 1s\n",
      "  7600K .......... .......... .......... .......... .......... 61% 2.14M 1s\n",
      "  7650K .......... .......... .......... .......... .......... 62% 8.70M 1s\n",
      "  7700K .......... .......... .......... .......... .......... 62% 21.8M 1s\n",
      "  7750K .......... .......... .......... .......... .......... 62% 28.8M 1s\n",
      "  7800K .......... .......... .......... .......... .......... 63% 8.57M 1s\n",
      "  7850K .......... .......... .......... .......... .......... 63% 97.1M 1s\n",
      "  7900K .......... .......... .......... .......... .......... 64% 4.45M 1s\n",
      "  7950K .......... .......... .......... .......... .......... 64% 1.65M 1s\n",
      "  8000K .......... .......... .......... .......... .......... 64% 13.5M 1s\n",
      "  8050K .......... .......... .......... .......... .......... 65%  543M 1s\n",
      "  8100K .......... .......... .......... .......... .......... 65%  488M 1s\n",
      "  8150K .......... .......... .......... .......... .......... 66% 42.8M 1s\n",
      "  8200K .......... .......... .......... .......... .......... 66% 94.4M 1s\n",
      "  8250K .......... .......... .......... .......... .......... 66%  407M 1s\n",
      "  8300K .......... .......... .......... .......... .......... 67%  381M 1s\n",
      "  8350K .......... .......... .......... .......... .......... 67%  317M 1s\n",
      "  8400K .......... .......... .......... .......... .......... 68%  452M 1s\n",
      "  8450K .......... .......... .......... .......... .......... 68%  421M 1s\n",
      "  8500K .......... .......... .......... .......... .......... 68%  428M 0s\n",
      "  8550K .......... .......... .......... .......... .......... 69%  102M 0s\n",
      "  8600K .......... .......... .......... .......... .......... 69%  444M 0s\n",
      "  8650K .......... .......... .......... .......... .......... 70%  514M 0s\n",
      "  8700K .......... .......... .......... .......... .......... 70%  531M 0s\n",
      "  8750K .......... .......... .......... .......... .......... 70%  595M 0s\n",
      "  8800K .......... .......... .......... .......... .......... 71%  581M 0s\n",
      "  8850K .......... .......... .......... .......... .......... 71%  543M 0s\n",
      "  8900K .......... .......... .......... .......... .......... 72%  581M 0s\n",
      "  8950K .......... .......... .......... .......... .......... 72%  370M 0s\n",
      "  9000K .......... .......... .......... .......... .......... 72%  568M 0s\n",
      "  9050K .......... .......... .......... .......... .......... 73%  595M 0s\n",
      "  9100K .......... .......... .......... .......... .......... 73%  417M 0s\n",
      "  9150K .......... .......... .......... .......... .......... 74% 80.4M 0s\n",
      "  9200K .......... .......... .......... .......... .......... 74% 1.17M 0s\n",
      "  9250K .......... .......... .......... .......... .......... 74% 1.60M 0s\n",
      "  9300K .......... .......... .......... .......... .......... 75% 8.64M 0s\n",
      "  9350K .......... .......... .......... .......... .......... 75% 2.75M 0s\n",
      "  9400K .......... .......... .......... .......... .......... 76%  561M 0s\n",
      "  9450K .......... .......... .......... .......... .......... 76%  300M 0s\n",
      "  9500K .......... .......... .......... .......... .......... 76%  488M 0s\n",
      "  9550K .......... .......... .......... .......... .......... 77%  162M 0s\n",
      "  9600K .......... .......... .......... .......... .......... 77% 85.1M 0s\n",
      "  9650K .......... .......... .......... .......... .......... 78%  509M 0s\n",
      "  9700K .......... .......... .......... .......... .......... 78%  525M 0s\n",
      "  9750K .......... .......... .......... .......... .......... 78%  543M 0s\n",
      "  9800K .......... .......... .......... .......... .......... 79%  555M 0s\n",
      "  9850K .......... .......... .......... .......... .......... 79%  444M 0s\n",
      "  9900K .......... .......... .......... .......... .......... 80%  209M 0s\n",
      "  9950K .......... .......... .......... .......... .......... 80%  328M 0s\n",
      " 10000K .......... .......... .......... .......... .......... 80%  479M 0s\n",
      " 10050K .......... .......... .......... .......... .......... 81%  549M 0s\n",
      " 10100K .......... .......... .......... .......... .......... 81%  483M 0s\n",
      " 10150K .......... .......... .......... .......... .......... 82%  362M 0s\n",
      " 10200K .......... .......... .......... .......... .......... 82% 3.25M 0s\n",
      " 10250K .......... .......... .......... .......... .......... 82% 20.3M 0s\n",
      " 10300K .......... .......... .......... .......... .......... 83%  595M 0s\n",
      " 10350K .......... .......... .......... .......... .......... 83%  514M 0s\n",
      " 10400K .......... .......... .......... .......... .......... 84% 27.0M 0s\n",
      " 10450K .......... .......... .......... .......... .......... 84% 25.6M 0s\n",
      " 10500K .......... .......... .......... .......... .......... 84% 16.3M 0s\n",
      " 10550K .......... .......... .......... .......... .......... 85% 29.7M 0s\n",
      " 10600K .......... .......... .......... .......... .......... 85%  334M 0s\n",
      " 10650K .......... .......... .......... .......... .......... 86%  421M 0s\n",
      " 10700K .......... .......... .......... .......... .......... 86% 99.4M 0s\n",
      " 10750K .......... .......... .......... .......... .......... 87%  461M 0s\n",
      " 10800K .......... .......... .......... .......... .......... 87% 40.1M 0s\n",
      " 10850K .......... .......... .......... .......... .......... 87% 8.80M 0s\n",
      " 10900K .......... .......... .......... .......... .......... 88%  407M 0s\n",
      " 10950K .......... .......... .......... .......... .......... 88%  376M 0s\n",
      " 11000K .......... .......... .......... .......... .......... 89%  588M 0s\n",
      " 11050K .......... .......... .......... .......... .......... 89%  595M 0s\n",
      " 11100K .......... .......... .......... .......... .......... 89%  112M 0s\n",
      " 11150K .......... .......... .......... .......... .......... 90% 7.34M 0s\n",
      " 11200K .......... .......... .......... .......... .......... 90%  111M 0s\n",
      " 11250K .......... .......... .......... .......... .......... 91%  330M 0s\n",
      " 11300K .......... .......... .......... .......... .......... 91% 31.2M 0s\n",
      " 11350K .......... .......... .......... .......... .......... 91%  351M 0s\n",
      " 11400K .......... .......... .......... .......... .......... 92%  531M 0s\n",
      " 11450K .......... .......... .......... .......... .......... 92%  356M 0s\n",
      " 11500K .......... .......... .......... .......... .......... 93%  531M 0s\n",
      " 11550K .......... .......... .......... .......... .......... 93%  309M 0s\n",
      " 11600K .......... .......... .......... .......... .......... 93%  391M 0s\n",
      " 11650K .......... .......... .......... .......... .......... 94%  493M 0s\n",
      " 11700K .......... .......... .......... .......... .......... 94%  488M 0s\n",
      " 11750K .......... .......... .......... .......... .......... 95% 31.3M 0s\n",
      " 11800K .......... .......... .......... .......... .......... 95% 23.5M 0s\n",
      " 11850K .......... .......... .......... .......... .......... 95%  311M 0s\n",
      " 11900K .......... .......... .......... .......... .......... 96%  595M 0s\n",
      " 11950K .......... .......... .......... .......... .......... 96%  555M 0s\n",
      " 12000K .......... .......... .......... .......... .......... 97%  370M 0s\n",
      " 12050K .......... .......... .......... .......... .......... 97%  102M 0s\n",
      " 12100K .......... .......... .......... .......... .......... 97% 53.3M 0s\n",
      " 12150K .......... .......... .......... .......... .......... 98% 5.60M 0s\n",
      " 12200K .......... .......... .......... .......... .......... 98% 28.9M 0s\n",
      " 12250K .......... .......... .......... .......... .......... 99% 15.7M 0s\n",
      " 12300K .......... .......... .......... .......... .......... 99% 40.8M 0s\n",
      " 12350K .......... .......... .......... .......... .......... 99% 12.4M 0s\n",
      " 12400K .......... .                                          100% 11.1M=1.3s\n",
      "\n",
      "2025-08-11 21:33:57 (9.52 MB/s) - ‘clones/chr21.fa.gz’ saved [12709705/12709705]\n",
      "\n"
     ]
    }
   ],
   "execution_count": 69
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:58.900867Z",
     "start_time": "2025-08-12T01:33:58.699980Z"
    }
   },
   "source": [
    "%%sh\n",
    "# copy the YAML config file\n",
    "cp ./configs/clones.yaml ./clones/.\n",
    "\n",
    "# display the config\n",
    "cat ./clones/clones.yaml"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "reference: ./clones/chr21.fa\n",
      "config_files:\n",
      "                0: 'configs/clone_configs/svs_0.yaml'\n",
      "                1: 'configs/clone_configs/svs_1.yaml'\n",
      "                2: 'configs/clone_configs/svs_2.yaml'\n",
      "                3: 'configs/clone_configs/svs_3.yaml'\n",
      "                4: 'configs/clone_configs/svs_4.yaml'\n",
      "\n",
      "\n",
      "clones:\n",
      "    A:\n",
      "     - 0\n",
      "    B:\n",
      "      - 0\n",
      "      - 1\n",
      "    C:\n",
      "      - 0\n",
      "      - 1\n",
      "      - 2\n",
      "    D:\n",
      "      - 0\n",
      "      - 1\n",
      "      - 3\n",
      "    E: [0, 4]\n",
      "\n",
      "purity:\n",
      "  A: 30\n",
      "  B: 20\n",
      "  C: 20\n",
      "  D: 20\n",
      "  E: 10\n",
      "\n",
      "coverage: 30"
     ]
    }
   ],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T14:39:47.218853Z",
     "start_time": "2025-08-12T14:39:47.206680Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def run_bash_process(command):\n",
    "    try:\n",
    "        output = subprocess.run(\n",
    "        ' '.join(command),\n",
    "        stdout=subprocess.DEVNULL,\n",
    "        stderr=subprocess.PIPE,\n",
    "        shell=True,\n",
    "        text=True          \n",
    "        )\n",
    "        \n",
    "        if output.stderr:\n",
    "            print(\"STDERR:\", output.stderr)\n",
    "\n",
    "    except subprocess.CalledProcessError as e:\n",
    "        print(\"Command failed with return code:\", e.returncode)\n",
    "        print(e.stderr)\n",
    "        raise        "
   ],
   "outputs": [],
   "execution_count": 93
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:58.939602Z",
     "start_time": "2025-08-12T01:33:58.935192Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def call_insilicosv(path, config_name):\n",
    "    command = ['insilicosv', '-c', path + config_name]\n",
    "    run_bash_process(command)\n",
    "    \n",
    "    merge_command = ['cat', path + 'sim.hapA.fa',  path + 'sim.hapB.fa', '>', path + 'sim.fa']\n",
    "    run_bash_process(merge_command)\n"
   ],
   "outputs": [],
   "execution_count": 72
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:33:58.972076Z",
     "start_time": "2025-08-12T01:33:58.958008Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def clonal_genome_generator(folder_path, config_name):\n",
    "    reads_to_simulate = {}\n",
    "    vcf_list = []\n",
    "    with open(folder_path + config_name) as config_yaml:\n",
    "        config = yaml.safe_load(config_yaml)\n",
    "\n",
    "    for clone_name, dependencies in config['clones'].items():\n",
    "        current_path = folder_path \n",
    "        previous_path = ''\n",
    "        lineage_depth = 0\n",
    "        for dependency in dependencies:\n",
    "            lineage_depth += 1\n",
    "            current_path = current_path + '/dependency_' + str(dependency) + '/'\n",
    "            config_name = config['config_files'][dependency].split('/')[-1]\n",
    "\n",
    "            if os.path.exists(current_path): \n",
    "                previous_path = current_path + '/sim.fa' \n",
    "                continue\n",
    "            \n",
    "            os.makedirs(current_path)\n",
    "            shutil.copy(config['config_files'][dependency], current_path)\n",
    "            \n",
    "            # Append the previous insilicoSV output as new reference or use the initial reference if first call\n",
    "            reference = config['reference']\n",
    "            if previous_path:\n",
    "                reference = previous_path\n",
    "            with open(current_path + config_name, 'a') as file:\n",
    "                file.write('\\nreference: ' + reference + \"\\n\")\n",
    "            call_insilicosv(current_path, config_name)\n",
    "            previous_path = current_path + '/sim.fa'\n",
    "            vcf_list.append(current_path + 'sim.vcf')\n",
    "        purity = config['purity'][clone_name]\n",
    "        reads_to_simulate[clone_name] = [previous_path.split('sim.fa')[0], purity, lineage_depth]\n",
    "    global_coverage = config['coverage']\n",
    "    reference = config['reference']\n",
    "    return reference, global_coverage, vcf_list, reads_to_simulate"
   ],
   "outputs": [],
   "execution_count": 73
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:34:07.140678Z",
     "start_time": "2025-08-12T01:33:58.990801Z"
    }
   },
   "cell_type": "code",
   "source": [
    "root_path = './clones/'\n",
    "reference, coverage, vcf_list, reads_to_simulate = clonal_genome_generator(root_path, 'clones.yaml')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "STDERR: 2025-08-11 21:34:00,798 INFO insilicoSV version v0.1.3\n",
      "2025-08-11 21:34:01,096 INFO Constructing SVs from 5 categories\n",
      "2025-08-11 21:34:01,099 INFO Constructed 25 SVs\n",
      "2025-08-11 21:34:01,099 INFO Deciding placement order for 25 SVs\n",
      "2025-08-11 21:34:01,099 INFO Placing 25 svs\n",
      "2025-08-11 21:34:01,106 INFO Placed 25 svs in 0.0s.\n",
      "2025-08-11 21:34:01,106 INFO Writing outputs\n",
      "2025-08-11 21:34:01,106 INFO Writing new haplotypes\n",
      "2025-08-11 21:34:01,288 INFO Writing VCF file\n",
      "2025-08-11 21:34:01,313 INFO Writing novel insertions file\n",
      "2025-08-11 21:34:01,313 INFO Writing novel adjacencies file\n",
      "2025-08-11 21:34:01,313 WARNING Skipping adjacencies output\n",
      "2025-08-11 21:34:01,313 INFO Output path: ./clones//dependency_0\n",
      "2025-08-11 21:34:01,313 INFO insilicoSV finished in 0.5s\n",
      "\n",
      "STDERR: 2025-08-11 21:34:01,906 INFO insilicoSV version v0.1.3\n",
      "2025-08-11 21:34:02,453 INFO Constructing SVs from 1 categories\n",
      "2025-08-11 21:34:02,456 INFO Constructed 50 SVs\n",
      "2025-08-11 21:34:02,457 INFO Deciding placement order for 50 SVs\n",
      "2025-08-11 21:34:02,457 INFO Placing 50 svs\n",
      "2025-08-11 21:34:02,465 INFO Placed 50 svs in 0.0s.\n",
      "2025-08-11 21:34:02,465 INFO Writing outputs\n",
      "2025-08-11 21:34:02,465 INFO Writing new haplotypes\n",
      "2025-08-11 21:34:02,623 INFO Writing VCF file\n",
      "2025-08-11 21:34:02,629 INFO Writing novel insertions file\n",
      "2025-08-11 21:34:02,629 INFO Writing novel adjacencies file\n",
      "2025-08-11 21:34:02,629 WARNING Skipping adjacencies output\n",
      "2025-08-11 21:34:02,630 INFO Output path: ./clones//dependency_0//dependency_1\n",
      "2025-08-11 21:34:02,630 INFO insilicoSV finished in 0.7s\n",
      "\n",
      "STDERR: 2025-08-11 21:34:03,399 INFO insilicoSV version v0.1.3\n",
      "2025-08-11 21:34:04,508 INFO Constructing SVs from 3 categories\n",
      "2025-08-11 21:34:04,513 INFO Constructed 60 SVs\n",
      "2025-08-11 21:34:04,513 INFO Deciding placement order for 60 SVs\n",
      "2025-08-11 21:34:04,513 INFO Placing 60 svs\n",
      "2025-08-11 21:34:04,526 INFO Placed 60 svs in 0.0s.\n",
      "2025-08-11 21:34:04,526 INFO Writing outputs\n",
      "2025-08-11 21:34:04,526 INFO Writing new haplotypes\n",
      "2025-08-11 21:34:04,858 INFO Writing VCF file\n",
      "2025-08-11 21:34:04,863 INFO Writing novel insertions file\n",
      "2025-08-11 21:34:04,863 INFO Writing novel adjacencies file\n",
      "2025-08-11 21:34:04,863 WARNING Skipping adjacencies output\n",
      "2025-08-11 21:34:04,864 INFO Output path: ./clones//dependency_0//dependency_1//dependency_2\n",
      "2025-08-11 21:34:04,864 INFO insilicoSV finished in 1.5s\n",
      "\n",
      "STDERR: 2025-08-11 21:34:05,577 INFO insilicoSV version v0.1.3\n",
      "2025-08-11 21:34:05,582 INFO Constructing SVs from 1 categories\n",
      "2025-08-11 21:34:05,582 INFO Constructed 5 SVs\n",
      "2025-08-11 21:34:05,583 INFO Deciding placement order for 5 SVs\n",
      "2025-08-11 21:34:05,583 INFO Placing 5 svs\n",
      "2025-08-11 21:34:05,585 INFO Placed 5 svs in 0.0s.\n",
      "2025-08-11 21:34:05,585 INFO Writing outputs\n",
      "2025-08-11 21:34:05,585 INFO Writing new haplotypes\n",
      "2025-08-11 21:34:05,883 INFO Writing VCF file\n",
      "2025-08-11 21:34:05,886 INFO Writing novel insertions file\n",
      "2025-08-11 21:34:05,886 INFO Writing novel adjacencies file\n",
      "2025-08-11 21:34:05,887 WARNING Skipping adjacencies output\n",
      "2025-08-11 21:34:05,887 INFO Output path: ./clones//dependency_0//dependency_1//dependency_3\n",
      "2025-08-11 21:34:05,887 INFO insilicoSV finished in 0.3s\n",
      "\n",
      "STDERR: 2025-08-11 21:34:06,731 INFO insilicoSV version v0.1.3\n",
      "2025-08-11 21:34:06,738 INFO Constructing SVs from 1 categories\n",
      "2025-08-11 21:34:06,739 INFO Constructed 5 SVs\n",
      "2025-08-11 21:34:06,739 INFO Deciding placement order for 5 SVs\n",
      "2025-08-11 21:34:06,739 INFO Placing 5 svs\n",
      "2025-08-11 21:34:06,741 INFO Placed 5 svs in 0.0s.\n",
      "2025-08-11 21:34:06,741 INFO Writing outputs\n",
      "2025-08-11 21:34:06,741 INFO Writing new haplotypes\n",
      "2025-08-11 21:34:06,979 INFO Writing VCF file\n",
      "2025-08-11 21:34:06,984 INFO Writing novel insertions file\n",
      "2025-08-11 21:34:06,984 INFO Writing novel adjacencies file\n",
      "2025-08-11 21:34:06,984 WARNING Skipping adjacencies output\n",
      "2025-08-11 21:34:06,984 INFO Output path: ./clones//dependency_0//dependency_4\n",
      "2025-08-11 21:34:06,984 INFO insilicoSV finished in 0.3s\n",
      "\n"
     ]
    }
   ],
   "execution_count": 74
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## Read simulation\n",
    "### Short-read simulation\n",
    "Below we simulate paired-end short reads at the suited coverage given each clone purity and the requested total coverage using ```DWGSIM```. \n",
    "\n",
    "After generating the reads, we align them with minimap2 (short-read mode) and sort the alignments using ```samtools```."
   ]
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:34:07.165073Z",
     "start_time": "2025-08-12T01:34:07.153373Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def call_dwgsim(dwgsim_coverage, read_length, platform, genome, output_prefix):\n",
    "    dwgsim_path = shutil.which('dwgsim')\n",
    "\n",
    "    command = [dwgsim_path, '-C', str(dwgsim_coverage), '-1', str(read_length), '-2', str(read_length), '-o', platform, '-H', genome, output_prefix]\n",
    "    run_bash_process(command)"
   ],
   "outputs": [],
   "execution_count": 75
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T15:05:48.842981Z",
     "start_time": "2025-08-12T15:05:48.839515Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def align_reads(platform, reference, reads, output_name):\n",
    "    command_align = ['minimap2', '-t', number_threads, '-ax', platform, reference] + reads + ['|', 'samtools', 'sort', '-@', number_threads, '-o', output_name, '-']\n",
    "    run_bash_process(command_align)\n",
    "    \n",
    "    command_index = ['samtools', 'index', '-@', number_threads, output_name]\n",
    "    run_bash_process(command_index)\n",
    "    \n",
    "    command_get_coverage = ['samtools', 'coverage', output_name, '>', output_name + '.coverage']\n",
    "    run_bash_process(command_get_coverage)\n",
    "    \n",
    "    with open(output_name + '.coverage', 'r') as cov_file:\n",
    "        print('PROCESSED: Clone', output_name, 'coverage/n', cov_file.read())"
   ],
   "outputs": [],
   "execution_count": 97
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:34:07.202770Z",
     "start_time": "2025-08-12T01:34:07.199709Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def merge_clones(output_name, list_bams):\n",
    "    print('Merging clones...')\n",
    "    command = ['samtools', 'merge', '-@', number_threads, '-o', output_name] + list_bams\n",
    "    run_bash_process(command)\n",
    "    \n",
    "    print('Indexing reads...')\n",
    "    command_index = ['samtools', 'index', '-@', number_threads, output_name]\n",
    "    run_bash_process(command_index)\n",
    "    \n",
    "    print('Computing coverage...')\n",
    "    command_get_coverage = ['samtools', 'coverage', output_name, '>', output_name + '.coverage']\n",
    "    run_bash_process(command_get_coverage)\n",
    "    \n",
    "    with open(output_name + '.coverage', 'r') as cov_file:\n",
    "        print('PROCESSED: Whole cancer genome', output_name, 'coverage', cov_file.read())"
   ],
   "outputs": [],
   "execution_count": 77
  },
  {
   "metadata": {},
   "cell_type": "markdown",
   "source": "We then merge all the simulated reads to obtain the requested coverage."
  },
  {
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2025-08-12T15:05:48.771301Z",
     "start_time": "2025-08-12T14:40:21.189595Z"
    }
   },
   "cell_type": "code",
   "source": [
    "read_length = 151\n",
    "dwgsim_platform = '0' # Illumina\n",
    "alignment_platform = 'sr'\n",
    "list_sr_bams = []\n",
    "for clone_name, (genome_folder, purity, lineage_depth) in reads_to_simulate.items():\n",
    "    print('Simulating reads for clone', clone_name)\n",
    "    output_prefix = genome_folder + 'sim_sr.dwgsim'\n",
    "    clone_coverage = ceil(coverage * purity / 100 / (2*lineage_depth))\n",
    "    print('Clone purity', purity, ' Clone coverage', clone_coverage)\n",
    "    call_dwgsim(clone_coverage, read_length, dwgsim_platform, genome_folder + '/sim.fa', output_prefix)\n",
    "    \n",
    "    print('Aligning reads')\n",
    "    r1 = output_prefix + '.bwa.read1.fastq.gz'\n",
    "    r2 = output_prefix + '.bwa.read2.fastq.gz'\n",
    "    output_name = genome_folder + '/' + clone_name + '_sim_sr.dwgsim.bam'\n",
    "    list_sr_bams.append(output_name)\n",
    "    reads = [r1, r2]\n",
    "    align_reads(alignment_platform, reference, reads, output_name)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulating reads for clone A\n",
      "Clone purity 30  Clone coverage 5\n",
      "STDERR: [dwgsim_core] chr21_hapA length: 46824501\n",
      "[dwgsim_core] chr21_hapB length: 46818423\n",
      "[dwgsim_core] 2 sequences, total length: 93642924\n",
      "[dwgsim_core] Currently on: \n",
      "0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 10000\n",
      "[dwgsim_core] 20000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 70000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 90000\n",
      "[dwgsim_core] 100000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 120000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 150000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 180000\n",
      "[dwgsim_core] 190000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 210000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 230000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 250000\n",
      "[dwgsim_core] 260000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 280000\n",
      "[dwgsim_core] 290000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 310000\n",
      "[dwgsim_core] 320000\n",
      "[dwgsim_core] 330000\n",
      "[dwgsim_core] 340000\n",
      "[dwgsim_core] 350000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 370000\n",
      "[dwgsim_core] 380000\n",
      "[dwgsim_core] 390000\n",
      "[dwgsim_core] 400000\n",
      "[dwgsim_core] 410000\n",
      "[dwgsim_core] 420000\n",
      "[dwgsim_core] 430000\n",
      "[dwgsim_core] 440000\n",
      "[dwgsim_core] 450000\n",
      "[dwgsim_core] 460000\n",
      "[dwgsim_core] 470000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 500000\n",
      "[dwgsim_core] 510000\n",
      "[dwgsim_core] 520000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 550000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 590000\n",
      "[dwgsim_core] 600000\n",
      "[dwgsim_core] 610000\n",
      "[dwgsim_core] 620000\n",
      "[dwgsim_core] 630000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 660000\n",
      "[dwgsim_core] 670000\n",
      "[dwgsim_core] 680000\n",
      "[dwgsim_core] 690000\n",
      "[dwgsim_core] 700000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 730000\n",
      "[dwgsim_core] 730000\n",
      "[dwgsim_core] 740000\n",
      "[dwgsim_core] 750000\n",
      "[dwgsim_core] 760000\n",
      "[dwgsim_core] 770000\n",
      "[dwgsim_core] 780000\n",
      "[dwgsim_core] 790000\n",
      "[dwgsim_core] 800000\n",
      "[dwgsim_core] 810000\n",
      "[dwgsim_core] 816042\n",
      "[dwgsim_core] 820000\n",
      "[dwgsim_core] 830000\n",
      "[dwgsim_core] 840000\n",
      "[dwgsim_core] 850000\n",
      "[dwgsim_core] 860000\n",
      "[dwgsim_core] 870000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 890000\n",
      "[dwgsim_core] 900000\n",
      "[dwgsim_core] 910000\n",
      "[dwgsim_core] 920000\n",
      "[dwgsim_core] 930000\n",
      "[dwgsim_core] 940000\n",
      "[dwgsim_core] 950000\n",
      "[dwgsim_core] 960000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 980000\n",
      "[dwgsim_core] 980000\n",
      "[dwgsim_core] 990000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1010000\n",
      "[dwgsim_core] 1020000\n",
      "[dwgsim_core] 1030000\n",
      "[dwgsim_core] 1040000\n",
      "[dwgsim_core] 1050000\n",
      "[dwgsim_core] 1060000\n",
      "[dwgsim_core] 1070000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1090000\n",
      "[dwgsim_core] 1100000\n",
      "[dwgsim_core] 1110000\n",
      "[dwgsim_core] 1120000\n",
      "[dwgsim_core] 1130000\n",
      "[dwgsim_core] 1140000\n",
      "[dwgsim_core] 1150000\n",
      "[dwgsim_core] 1160000\n",
      "[dwgsim_core] 1170000\n",
      "[dwgsim_core] 1180000\n",
      "[dwgsim_core] 1190000\n",
      "[dwgsim_core] 1200000\n",
      "[dwgsim_core] 1210000\n",
      "[dwgsim_core] 1220000\n",
      "[dwgsim_core] 1230000\n",
      "[dwgsim_core] 1240000\n",
      "[dwgsim_core] 1250000\n",
      "[dwgsim_core] 1260000\n",
      "[dwgsim_core] 1270000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1300000\n",
      "[dwgsim_core] 1310000\n",
      "[dwgsim_core] 1320000\n",
      "[dwgsim_core] 1330000\n",
      "[dwgsim_core] 1340000\n",
      "[dwgsim_core] 1350000\n",
      "[dwgsim_core] 1360000\n",
      "[dwgsim_core] 1370000\n",
      "[dwgsim_core] 1380000\n",
      "[dwgsim_core] 1390000\n",
      "[dwgsim_core] 1400000\n",
      "[dwgsim_core] 1410000\n",
      "[dwgsim_core] 1420000\n",
      "[dwgsim_core] 1430000\n",
      "[dwgsim_core] 1440000\n",
      "[dwgsim_core] 1450000\n",
      "[dwgsim_core] 1460000\n",
      "[dwgsim_core] 1470000\n",
      "[dwgsim_core] 1480000\n",
      "[dwgsim_core] 1490000\n",
      "[dwgsim_core] 1500000\n",
      "[dwgsim_core] 1510000\n",
      "[dwgsim_core] 1520000\n",
      "[dwgsim_core] 1530000\n",
      "[dwgsim_core] 1540000\n",
      "[dwgsim_core] 1550000\n",
      "[dwgsim_core] 1560000\n",
      "[dwgsim_core] 1570000\n",
      "[dwgsim_core] 1580000\n",
      "[dwgsim_core] 1590000\n",
      "[dwgsim_core] 1600000\n",
      "[dwgsim_core] 1610000\n",
      "[dwgsim_core] 1620000\n",
      "[dwgsim_core] 1630000\n",
      "[dwgsim_core] 1631978\n",
      "[dwgsim_core] Complete!\n",
      "\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.892*0.97] collected minimizers\n",
      "[M::mm_idx_gen::0.956*1.30] sorted minimizers\n",
      "[M::main::0.957*1.29] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.957*1.29] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.992*1.28] distinct minimizers: 5499655 (94.10% are singletons); average occurrences: 1.217; average spacing: 6.980; total length: 46709983\n",
      "[M::worker_pipeline::6.083*5.10] mapped 331126 sequences\n",
      "[M::worker_pipeline::9.578*5.59] mapped 331126 sequences\n",
      "[M::worker_pipeline::12.846*5.89] mapped 331126 sequences\n",
      "[M::worker_pipeline::17.448*5.67] mapped 331126 sequences\n",
      "[M::worker_pipeline::21.218*5.72] mapped 331126 sequences\n",
      "[M::worker_pipeline::24.674*5.84] mapped 331126 sequences\n",
      "[M::worker_pipeline::27.919*5.96] mapped 331126 sequences\n",
      "[M::worker_pipeline::31.388*5.99] mapped 331126 sequences\n",
      "[M::worker_pipeline::35.505*5.92] mapped 331126 sequences\n",
      "[M::worker_pipeline::38.140*5.86] mapped 283822 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax sr ./clones/chr21.fa ./clones//dependency_0//sim_sr.dwgsim.bwa.read1.fastq.gz ./clones//dependency_0//sim_sr.dwgsim.bwa.read2.fastq.gz\n",
      "[M::main] Real time: 38.164 sec; CPU: 223.691 sec; Peak RSS: 0.957 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0///A_sim_sr.dwgsim.bam > ./clones//dependency_0///A_sim_sr.dwgsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0///A_sim_sr.dwgsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t3075284\t40067452\t85.7792\t9.93644\t17.4\t54.4\n",
      "\n",
      "Simulating reads for clone B\n",
      "Clone purity 20  Clone coverage 2\n",
      "STDERR: [dwgsim_core] chr21_hapA_hapA length: 46824613\n",
      "[dwgsim_core] chr21_hapB_hapA length: 46818308\n",
      "[dwgsim_core] chr21_hapA_hapB length: 46824817\n",
      "[dwgsim_core] chr21_hapB_hapB length: 46818182\n",
      "[dwgsim_core] 4 sequences, total length: 187285920\n",
      "[dwgsim_core] Currently on: \n",
      "0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 10000\n",
      "[dwgsim_core] 20000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 70000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 90000\n",
      "[dwgsim_core] 100000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 120000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 150000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 180000\n",
      "[dwgsim_core] 190000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 210000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 230000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 250000\n",
      "[dwgsim_core] 260000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 280000\n",
      "[dwgsim_core] 290000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 310000\n",
      "[dwgsim_core] 320000\n",
      "[dwgsim_core] 326418\n",
      "[dwgsim_core] 330000\n",
      "[dwgsim_core] 340000\n",
      "[dwgsim_core] 350000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 370000\n",
      "[dwgsim_core] 380000\n",
      "[dwgsim_core] 390000\n",
      "[dwgsim_core] 400000\n",
      "[dwgsim_core] 410000\n",
      "[dwgsim_core] 420000\n",
      "[dwgsim_core] 430000\n",
      "[dwgsim_core] 440000\n",
      "[dwgsim_core] 450000\n",
      "[dwgsim_core] 460000\n",
      "[dwgsim_core] 470000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 500000\n",
      "[dwgsim_core] 510000\n",
      "[dwgsim_core] 520000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 550000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 590000\n",
      "[dwgsim_core] 600000\n",
      "[dwgsim_core] 610000\n",
      "[dwgsim_core] 620000\n",
      "[dwgsim_core] 630000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 652792\n",
      "[dwgsim_core] 660000\n",
      "[dwgsim_core] 670000\n",
      "[dwgsim_core] 680000\n",
      "[dwgsim_core] 690000\n",
      "[dwgsim_core] 690000\n",
      "[dwgsim_core] 700000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 730000\n",
      "[dwgsim_core] 740000\n",
      "[dwgsim_core] 750000\n",
      "[dwgsim_core] 760000\n",
      "[dwgsim_core] 770000\n",
      "[dwgsim_core] 780000\n",
      "[dwgsim_core] 790000\n",
      "[dwgsim_core] 800000\n",
      "[dwgsim_core] 810000\n",
      "[dwgsim_core] 820000\n",
      "[dwgsim_core] 830000\n",
      "[dwgsim_core] 840000\n",
      "[dwgsim_core] 850000\n",
      "[dwgsim_core] 860000\n",
      "[dwgsim_core] 870000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 890000\n",
      "[dwgsim_core] 900000\n",
      "[dwgsim_core] 910000\n",
      "[dwgsim_core] 920000\n",
      "[dwgsim_core] 930000\n",
      "[dwgsim_core] 940000\n",
      "[dwgsim_core] 950000\n",
      "[dwgsim_core] 960000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 979211\n",
      "[dwgsim_core] 980000\n",
      "[dwgsim_core] 990000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1010000\n",
      "[dwgsim_core] 1020000\n",
      "[dwgsim_core] 1030000\n",
      "[dwgsim_core] 1040000\n",
      "[dwgsim_core] 1050000\n",
      "[dwgsim_core] 1060000\n",
      "[dwgsim_core] 1060000\n",
      "[dwgsim_core] 1070000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1090000\n",
      "[dwgsim_core] 1100000\n",
      "[dwgsim_core] 1110000\n",
      "[dwgsim_core] 1120000\n",
      "[dwgsim_core] 1130000\n",
      "[dwgsim_core] 1140000\n",
      "[dwgsim_core] 1150000\n",
      "[dwgsim_core] 1160000\n",
      "[dwgsim_core] 1170000\n",
      "[dwgsim_core] 1180000\n",
      "[dwgsim_core] 1190000\n",
      "[dwgsim_core] 1200000\n",
      "[dwgsim_core] 1210000\n",
      "[dwgsim_core] 1220000\n",
      "[dwgsim_core] 1230000\n",
      "[dwgsim_core] 1240000\n",
      "[dwgsim_core] 1250000\n",
      "[dwgsim_core] 1260000\n",
      "[dwgsim_core] 1270000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1300000\n",
      "[dwgsim_core] 1305584\n",
      "[dwgsim_core] Complete!\n",
      "\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.813*1.00] collected minimizers\n",
      "[M::mm_idx_gen::0.877*1.36] sorted minimizers\n",
      "[M::main::0.878*1.36] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.878*1.36] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.915*1.34] distinct minimizers: 5499655 (94.10% are singletons); average occurrences: 1.217; average spacing: 6.980; total length: 46709983\n",
      "[M::worker_pipeline::5.785*5.50] mapped 331126 sequences\n",
      "[M::worker_pipeline::9.028*5.92] mapped 331126 sequences\n",
      "[M::worker_pipeline::13.365*5.78] mapped 331126 sequences\n",
      "[M::worker_pipeline::16.549*6.02] mapped 331126 sequences\n",
      "[M::worker_pipeline::19.787*6.19] mapped 331126 sequences\n",
      "[M::worker_pipeline::22.787*6.35] mapped 331126 sequences\n",
      "[M::worker_pipeline::26.028*6.43] mapped 331126 sequences\n",
      "[M::worker_pipeline::28.015*6.43] mapped 293286 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax sr ./clones/chr21.fa ./clones//dependency_0//dependency_1//sim_sr.dwgsim.bwa.read1.fastq.gz ./clones//dependency_0//dependency_1//sim_sr.dwgsim.bwa.read2.fastq.gz\n",
      "[M::main] Real time: 28.030 sec; CPU: 180.029 sec; Peak RSS: 1.124 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1///B_sim_sr.dwgsim.bam > ./clones//dependency_0//dependency_1///B_sim_sr.dwgsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1///B_sim_sr.dwgsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t2460835\t40062559\t85.7687\t7.95096\t17.4\t54.4\n",
      "\n",
      "Simulating reads for clone C\n",
      "Clone purity 20  Clone coverage 1\n",
      "STDERR: [dwgsim_core] chr21_hapA_hapA_hapA length: 46807604\n",
      "[dwgsim_core] chr21_hapB_hapA_hapA length: 46816024\n",
      "[dwgsim_core] chr21_hapA_hapB_hapA length: 46831148\n",
      "[dwgsim_core] chr21_hapB_hapB_hapA length: 46809734\n",
      "[dwgsim_core] chr21_hapA_hapA_hapB length: 46828375\n",
      "[dwgsim_core] chr21_hapB_hapA_hapB length: 46822204\n",
      "[dwgsim_core] chr21_hapA_hapB_hapB length: 46822836\n",
      "[dwgsim_core] chr21_hapB_hapB_hapB length: 46809734\n",
      "[dwgsim_core] 8 sequences, total length: 374547659\n",
      "[dwgsim_core] Currently on: \n",
      "0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 10000\n",
      "[dwgsim_core] 20000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 70000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 90000\n",
      "[dwgsim_core] 100000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 120000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 150000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 163150\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 180000\n",
      "[dwgsim_core] 190000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 210000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 230000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 250000\n",
      "[dwgsim_core] 260000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 280000\n",
      "[dwgsim_core] 290000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 310000\n",
      "[dwgsim_core] 320000\n",
      "[dwgsim_core] 326329\n",
      "[dwgsim_core] 330000\n",
      "[dwgsim_core] 340000\n",
      "[dwgsim_core] 350000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 370000\n",
      "[dwgsim_core] 380000\n",
      "[dwgsim_core] 390000\n",
      "[dwgsim_core] 400000\n",
      "[dwgsim_core] 410000\n",
      "[dwgsim_core] 420000\n",
      "[dwgsim_core] 430000\n",
      "[dwgsim_core] 440000\n",
      "[dwgsim_core] 450000\n",
      "[dwgsim_core] 460000\n",
      "[dwgsim_core] 470000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 489561\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 500000\n",
      "[dwgsim_core] 510000\n",
      "[dwgsim_core] 520000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 550000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 590000\n",
      "[dwgsim_core] 600000\n",
      "[dwgsim_core] 610000\n",
      "[dwgsim_core] 620000\n",
      "[dwgsim_core] 630000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 652718\n",
      "[dwgsim_core] 660000\n",
      "[dwgsim_core] 670000\n",
      "[dwgsim_core] 680000\n",
      "[dwgsim_core] 690000\n",
      "[dwgsim_core] 700000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 730000\n",
      "[dwgsim_core] 740000\n",
      "[dwgsim_core] 750000\n",
      "[dwgsim_core] 760000\n",
      "[dwgsim_core] 770000\n",
      "[dwgsim_core] 780000\n",
      "[dwgsim_core] 790000\n",
      "[dwgsim_core] 800000\n",
      "[dwgsim_core] 810000\n",
      "[dwgsim_core] 815940\n",
      "[dwgsim_core] 820000\n",
      "[dwgsim_core] 830000\n",
      "[dwgsim_core] 840000\n",
      "[dwgsim_core] 850000\n",
      "[dwgsim_core] 860000\n",
      "[dwgsim_core] 870000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 890000\n",
      "[dwgsim_core] 900000\n",
      "[dwgsim_core] 910000\n",
      "[dwgsim_core] 920000\n",
      "[dwgsim_core] 930000\n",
      "[dwgsim_core] 940000\n",
      "[dwgsim_core] 950000\n",
      "[dwgsim_core] 960000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 979140\n",
      "[dwgsim_core] 980000\n",
      "[dwgsim_core] 990000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1010000\n",
      "[dwgsim_core] 1020000\n",
      "[dwgsim_core] 1030000\n",
      "[dwgsim_core] 1040000\n",
      "[dwgsim_core] 1040000\n",
      "[dwgsim_core] 1050000\n",
      "[dwgsim_core] 1060000\n",
      "[dwgsim_core] 1070000\n",
      "[dwgsim_core] 1070000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1090000\n",
      "[dwgsim_core] 1100000\n",
      "[dwgsim_core] 1110000\n",
      "[dwgsim_core] 1120000\n",
      "[dwgsim_core] 1130000\n",
      "[dwgsim_core] 1140000\n",
      "[dwgsim_core] 1142343\n",
      "[dwgsim_core] 1150000\n",
      "[dwgsim_core] 1160000\n",
      "[dwgsim_core] 1170000\n",
      "[dwgsim_core] 1180000\n",
      "[dwgsim_core] 1180000\n",
      "[dwgsim_core] 1190000\n",
      "[dwgsim_core] 1200000\n",
      "[dwgsim_core] 1210000\n",
      "[dwgsim_core] 1220000\n",
      "[dwgsim_core] 1230000\n",
      "[dwgsim_core] 1240000\n",
      "[dwgsim_core] 1240000\n",
      "[dwgsim_core] 1250000\n",
      "[dwgsim_core] 1260000\n",
      "[dwgsim_core] 1270000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1300000\n",
      "[dwgsim_core] 1305500\n",
      "[dwgsim_core] Complete!\n",
      "\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.846*1.00] collected minimizers\n",
      "[M::mm_idx_gen::0.906*1.36] sorted minimizers\n",
      "[M::main::0.907*1.36] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.907*1.36] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.941*1.34] distinct minimizers: 5499655 (94.10% are singletons); average occurrences: 1.217; average spacing: 6.980; total length: 46709983\n",
      "[M::worker_pipeline::7.050*4.36] mapped 331126 sequences\n",
      "[M::worker_pipeline::11.166*4.80] mapped 331126 sequences\n",
      "[M::worker_pipeline::15.415*5.06] mapped 331126 sequences\n",
      "[M::worker_pipeline::18.935*5.31] mapped 331126 sequences\n",
      "[M::worker_pipeline::22.635*5.45] mapped 331126 sequences\n",
      "[M::worker_pipeline::25.755*5.64] mapped 331126 sequences\n",
      "[M::worker_pipeline::28.900*5.80] mapped 331126 sequences\n",
      "[M::worker_pipeline::30.847*5.84] mapped 293118 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax sr ./clones/chr21.fa ./clones//dependency_0//dependency_1//dependency_2//sim_sr.dwgsim.bwa.read1.fastq.gz ./clones//dependency_0//dependency_1//dependency_2//sim_sr.dwgsim.bwa.read2.fastq.gz\n",
      "[M::main] Real time: 30.864 sec; CPU: 180.129 sec; Peak RSS: 0.992 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1//dependency_2///C_sim_sr.dwgsim.bam > ./clones//dependency_0//dependency_1//dependency_2///C_sim_sr.dwgsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1//dependency_2///C_sim_sr.dwgsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t2460345\t40063279\t85.7703\t7.94942\t17.4\t54.5\n",
      "\n",
      "Simulating reads for clone D\n",
      "Clone purity 20  Clone coverage 1\n",
      "STDERR: [dwgsim_core] chr21_hapA_hapA_hapA length: 46830189\n",
      "[dwgsim_core] chr21_hapB_hapA_hapA length: 46818308\n",
      "[dwgsim_core] chr21_hapA_hapB_hapA length: 46824817\n",
      "[dwgsim_core] chr21_hapB_hapB_hapA length: 46824419\n",
      "[dwgsim_core] chr21_hapA_hapA_hapB length: 46824613\n",
      "[dwgsim_core] chr21_hapB_hapA_hapB length: 46826082\n",
      "[dwgsim_core] chr21_hapA_hapB_hapB length: 46824817\n",
      "[dwgsim_core] chr21_hapB_hapB_hapB length: 46833820\n",
      "[dwgsim_core] 8 sequences, total length: 374607065\n",
      "[dwgsim_core] Currently on: \n",
      "0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 10000\n",
      "[dwgsim_core] 20000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 70000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 90000\n",
      "[dwgsim_core] 100000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 120000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 150000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 163228\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 180000\n",
      "[dwgsim_core] 190000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 210000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 230000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 250000\n",
      "[dwgsim_core] 260000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 280000\n",
      "[dwgsim_core] 290000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 310000\n",
      "[dwgsim_core] 320000\n",
      "[dwgsim_core] 326415\n",
      "[dwgsim_core] 330000\n",
      "[dwgsim_core] 340000\n",
      "[dwgsim_core] 350000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 370000\n",
      "[dwgsim_core] 380000\n",
      "[dwgsim_core] 390000\n",
      "[dwgsim_core] 400000\n",
      "[dwgsim_core] 410000\n",
      "[dwgsim_core] 420000\n",
      "[dwgsim_core] 430000\n",
      "[dwgsim_core] 440000\n",
      "[dwgsim_core] 450000\n",
      "[dwgsim_core] 460000\n",
      "[dwgsim_core] 470000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 489625\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 500000\n",
      "[dwgsim_core] 510000\n",
      "[dwgsim_core] 520000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 550000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 590000\n",
      "[dwgsim_core] 600000\n",
      "[dwgsim_core] 610000\n",
      "[dwgsim_core] 620000\n",
      "[dwgsim_core] 630000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 652833\n",
      "[dwgsim_core] 660000\n",
      "[dwgsim_core] 670000\n",
      "[dwgsim_core] 680000\n",
      "[dwgsim_core] 680000\n",
      "[dwgsim_core] 690000\n",
      "[dwgsim_core] 700000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 710000\n",
      "[dwgsim_core] 720000\n",
      "[dwgsim_core] 730000\n",
      "[dwgsim_core] 740000\n",
      "[dwgsim_core] 740000\n",
      "[dwgsim_core] 750000\n",
      "[dwgsim_core] 760000\n",
      "[dwgsim_core] 770000\n",
      "[dwgsim_core] 780000\n",
      "[dwgsim_core] 790000\n",
      "[dwgsim_core] 800000\n",
      "[dwgsim_core] 810000\n",
      "[dwgsim_core] 816042\n",
      "[dwgsim_core] 820000\n",
      "[dwgsim_core] 830000\n",
      "[dwgsim_core] 830000\n",
      "[dwgsim_core] 840000\n",
      "[dwgsim_core] 850000\n",
      "[dwgsim_core] 850000\n",
      "[dwgsim_core] 860000\n",
      "[dwgsim_core] 870000\n",
      "[dwgsim_core] 880000\n",
      "[dwgsim_core] 890000\n",
      "[dwgsim_core] 900000\n",
      "[dwgsim_core] 910000\n",
      "[dwgsim_core] 920000\n",
      "[dwgsim_core] 930000\n",
      "[dwgsim_core] 940000\n",
      "[dwgsim_core] 950000\n",
      "[dwgsim_core] 960000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 970000\n",
      "[dwgsim_core] 979256\n",
      "[dwgsim_core] 980000\n",
      "[dwgsim_core] 990000\n",
      "[dwgsim_core] 1000000\n",
      "[dwgsim_core] 1010000\n",
      "[dwgsim_core] 1020000\n",
      "[dwgsim_core] 1030000\n",
      "[dwgsim_core] 1040000\n",
      "[dwgsim_core] 1050000\n",
      "[dwgsim_core] 1060000\n",
      "[dwgsim_core] 1070000\n",
      "[dwgsim_core] 1080000\n",
      "[dwgsim_core] 1090000\n",
      "[dwgsim_core] 1100000\n",
      "[dwgsim_core] 1110000\n",
      "[dwgsim_core] 1120000\n",
      "[dwgsim_core] 1130000\n",
      "[dwgsim_core] 1140000\n",
      "[dwgsim_core] 1142466\n",
      "[dwgsim_core] 1150000\n",
      "[dwgsim_core] 1150000\n",
      "[dwgsim_core] 1160000\n",
      "[dwgsim_core] 1170000\n",
      "[dwgsim_core] 1180000\n",
      "[dwgsim_core] 1190000\n",
      "[dwgsim_core] 1200000\n",
      "[dwgsim_core] 1210000\n",
      "[dwgsim_core] 1210000\n",
      "[dwgsim_core] 1220000\n",
      "[dwgsim_core] 1220000\n",
      "[dwgsim_core] 1230000\n",
      "[dwgsim_core] 1240000\n",
      "[dwgsim_core] 1250000\n",
      "[dwgsim_core] 1260000\n",
      "[dwgsim_core] 1270000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1280000\n",
      "[dwgsim_core] 1290000\n",
      "[dwgsim_core] 1300000\n",
      "[dwgsim_core] 1305707\n",
      "[dwgsim_core] Complete!\n",
      "\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.835*1.00] collected minimizers\n",
      "[M::mm_idx_gen::0.904*1.34] sorted minimizers\n",
      "[M::main::0.905*1.34] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.905*1.34] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.939*1.33] distinct minimizers: 5499655 (94.10% are singletons); average occurrences: 1.217; average spacing: 6.980; total length: 46709983\n",
      "[M::worker_pipeline::6.059*5.15] mapped 331126 sequences\n",
      "[M::worker_pipeline::9.422*5.79] mapped 331126 sequences\n",
      "[M::worker_pipeline::12.617*5.97] mapped 331126 sequences\n",
      "[M::worker_pipeline::16.256*6.05] mapped 331126 sequences\n",
      "[M::worker_pipeline::20.044*6.01] mapped 331126 sequences\n",
      "[M::worker_pipeline::23.832*6.01] mapped 331126 sequences\n",
      "[M::worker_pipeline::28.326*5.82] mapped 331126 sequences\n",
      "[M::worker_pipeline::30.655*5.85] mapped 293532 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax sr ./clones/chr21.fa ./clones//dependency_0//dependency_1//dependency_3//sim_sr.dwgsim.bwa.read1.fastq.gz ./clones//dependency_0//dependency_1//dependency_3//sim_sr.dwgsim.bwa.read2.fastq.gz\n",
      "[M::main] Real time: 30.670 sec; CPU: 179.497 sec; Peak RSS: 0.944 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1//dependency_3///D_sim_sr.dwgsim.bam > ./clones//dependency_0//dependency_1//dependency_3///D_sim_sr.dwgsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1//dependency_3///D_sim_sr.dwgsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t2459849\t40062273\t85.7681\t7.94784\t17.4\t54.5\n",
      "\n",
      "Simulating reads for clone E\n",
      "Clone purity 10  Clone coverage 1\n",
      "STDERR: [dwgsim_core] chr21_hapA_hapA length: 46824501\n",
      "[dwgsim_core] chr21_hapB_hapA length: 46818423\n",
      "[dwgsim_core] chr21_hapA_hapB length: 46824501\n",
      "[dwgsim_core] chr21_hapB_hapB length: 46818423\n",
      "[dwgsim_core] 4 sequences, total length: 187285848\n",
      "[dwgsim_core] Currently on: \n",
      "0\n",
      "[dwgsim_core] 0\n",
      "[dwgsim_core] 10000\n",
      "[dwgsim_core] 20000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 30000\n",
      "[dwgsim_core] 40000\n",
      "[dwgsim_core] 50000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 60000\n",
      "[dwgsim_core] 70000\n",
      "[dwgsim_core] 80000\n",
      "[dwgsim_core] 90000\n",
      "[dwgsim_core] 100000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 110000\n",
      "[dwgsim_core] 120000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 130000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 140000\n",
      "[dwgsim_core] 150000\n",
      "[dwgsim_core] 160000\n",
      "[dwgsim_core] 163208\n",
      "[dwgsim_core] 170000\n",
      "[dwgsim_core] 180000\n",
      "[dwgsim_core] 190000\n",
      "[dwgsim_core] 200000\n",
      "[dwgsim_core] 210000\n",
      "[dwgsim_core] 220000\n",
      "[dwgsim_core] 230000\n",
      "[dwgsim_core] 240000\n",
      "[dwgsim_core] 250000\n",
      "[dwgsim_core] 260000\n",
      "[dwgsim_core] 270000\n",
      "[dwgsim_core] 280000\n",
      "[dwgsim_core] 290000\n",
      "[dwgsim_core] 300000\n",
      "[dwgsim_core] 310000\n",
      "[dwgsim_core] 320000\n",
      "[dwgsim_core] 326395\n",
      "[dwgsim_core] 330000\n",
      "[dwgsim_core] 340000\n",
      "[dwgsim_core] 350000\n",
      "[dwgsim_core] 360000\n",
      "[dwgsim_core] 370000\n",
      "[dwgsim_core] 380000\n",
      "[dwgsim_core] 390000\n",
      "[dwgsim_core] 400000\n",
      "[dwgsim_core] 410000\n",
      "[dwgsim_core] 420000\n",
      "[dwgsim_core] 430000\n",
      "[dwgsim_core] 440000\n",
      "[dwgsim_core] 450000\n",
      "[dwgsim_core] 460000\n",
      "[dwgsim_core] 470000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 480000\n",
      "[dwgsim_core] 489603\n",
      "[dwgsim_core] 490000\n",
      "[dwgsim_core] 500000\n",
      "[dwgsim_core] 510000\n",
      "[dwgsim_core] 520000\n",
      "[dwgsim_core] 530000\n",
      "[dwgsim_core] 540000\n",
      "[dwgsim_core] 550000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 560000\n",
      "[dwgsim_core] 570000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 580000\n",
      "[dwgsim_core] 590000\n",
      "[dwgsim_core] 600000\n",
      "[dwgsim_core] 610000\n",
      "[dwgsim_core] 620000\n",
      "[dwgsim_core] 630000\n",
      "[dwgsim_core] 640000\n",
      "[dwgsim_core] 650000\n",
      "[dwgsim_core] 652790\n",
      "[dwgsim_core] Complete!\n",
      "\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.813*1.01] collected minimizers\n",
      "[M::mm_idx_gen::0.866*1.39] sorted minimizers\n",
      "[M::main::0.871*1.38] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.872*1.38] mid_occ = 1000\n",
      "[M::mm_idx_stat] kmer size: 21; skip: 11; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.906*1.37] distinct minimizers: 5499655 (94.10% are singletons); average occurrences: 1.217; average spacing: 6.980; total length: 46709983\n",
      "[M::worker_pipeline::6.009*5.20] mapped 331126 sequences\n",
      "[M::worker_pipeline::9.389*5.67] mapped 331126 sequences\n",
      "[M::worker_pipeline::12.634*5.90] mapped 331126 sequences\n",
      "[M::worker_pipeline::15.175*5.87] mapped 312202 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax sr ./clones/chr21.fa ./clones//dependency_0//dependency_4//sim_sr.dwgsim.bwa.read1.fastq.gz ./clones//dependency_0//dependency_4//sim_sr.dwgsim.bwa.read2.fastq.gz\n",
      "[M::main] Real time: 15.194 sec; CPU: 89.132 sec; Peak RSS: 0.938 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_4///E_sim_sr.dwgsim.bam > ./clones//dependency_0//dependency_4///E_sim_sr.dwgsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_4///E_sim_sr.dwgsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t1229256\t39662843\t84.913\t3.97174\t17.4\t54.5\n",
      "\n"
     ]
    }
   ],
   "execution_count": 94
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:58:33.013663Z",
     "start_time": "2025-08-12T01:57:52.537129Z"
    }
   },
   "cell_type": "code",
   "source": [
    "output_name_sr = root_path + 'cancer_genome_sr.bam'\n",
    "merge_clones(output_name_sr, list_sr_bams)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging clones...\n",
      "Indexing reads...\n",
      "Computing coverage...\n",
      "PROCESSED: Whole cancer genome ./clones/cancer_genome_sr.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t11682857\t40069331\t85.7832\t37.7478\t17.4\t54.4\n",
      "\n"
     ]
    }
   ],
   "execution_count": 79
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "### Long-read simulation\n",
    "We use PBSIM3 to simulate HiFi reads from the synthetic genome. Since PBSIM3 outputs reads for each reference contig, we also combine the reads from the two synthetic haplotypes into a single FASTQ file.\n",
    "\n",
    "After generating the reads, we align them with minimap2 (HiFi mode) and sort the alignments using ```samtools```."
   ]
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T01:58:33.033360Z",
     "start_time": "2025-08-12T01:58:33.028062Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def call_pbsim(pbsim_coverage, read_length_mean, accuracy_mean, genome, output_prefix):\n",
    "    conda_prefix = os.environ.get('CONDA_PREFIX')\n",
    "    command = ['pbsim', '--depth', str(pbsim_coverage), '--genome', genome, '--prefix', output_prefix, '--strategy', 'wgs', '--method', 'qshmm', '--qshmm', conda_prefix + '/data/QSHMM-RSII.model', '--length-mean', read_length_mean, '--accuracy-mean', accuracy_mean]\n",
    "    run_bash_process(command)\n",
    "    print('Merging haplotypes')\n",
    "    merge_command = ['zcat', output_prefix + '*.fq.gz', '>>', output_prefix + '.fastq']\n",
    "    run_bash_process(merge_command)"
   ],
   "outputs": [],
   "execution_count": 80
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-12T14:25:15.224808Z",
     "start_time": "2025-08-12T14:22:11.781112Z"
    }
   },
   "cell_type": "code",
   "source": [
    "read_length_mean = '20000'\n",
    "accuracy_mean = '0.999'\n",
    "alignment_platform = 'map-hifi'\n",
    "print('Simulating long reads at coverage', coverage)\n",
    "list_lr_bams = []\n",
    "for clone_name, (genome_folder, purity, lineage_depth) in reads_to_simulate.items():\n",
    "    print('Simulating reads for clone', clone_name)\n",
    "    clone_coverage = coverage * purity / 100 / (2*lineage_depth)\n",
    "    print('Clone purity', purity, 'Clone coverage', clone_coverage)\n",
    "    output_prefix = genome_folder + 'sim_lr.pbsim'\n",
    "    call_pbsim(clone_coverage, read_length_mean, accuracy_mean, genome_folder + '/sim.fa', output_prefix)\n",
    "    \n",
    "    print('Aligning reads')\n",
    "    reads = [output_prefix + '.fastq']\n",
    "    output_name = genome_folder + '/' + clone_name + '_sim_lr.pbsim.bam'\n",
    "    list_lr_bams.append(output_name)\n",
    "    align_reads(alignment_platform, reference, reads, output_name)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Simulating long reads at coverage 30\n",
      "Simulating reads for clone A\n",
      "Clone purity 30 Clone coverage 4.5 30 30 1\n",
      "STDERR: :::: Simulation parameters :::\n",
      "\n",
      "strategy : wgs\n",
      "method : qshmm\n",
      "qshmm : /Users/ebattist/anaconda3/envs/insilicosv-demo-env/data/QSHMM-RSII.model\n",
      "genome : ./clones//dependency_0///sim.fa\n",
      "prefix : ./clones//dependency_0//sim_lr.pbsim\n",
      "id-prefix : S\n",
      "depth : 4.500000\n",
      "length-mean : 20000.000000\n",
      "length-sd : 7000.000000\n",
      "length-min : 100\n",
      "length-max : 1000000\n",
      "difference-ratio : 6:55:39\n",
      "seed : 1755008532\n",
      "accuracy-mean : 0.990000\n",
      "pass_num : 1\n",
      "hp-del-bias : 1.000000\n",
      "\n",
      ":::: Reference stats ::::\n",
      "\n",
      "file name : ./clones//dependency_0///sim.fa\n",
      "\n",
      "ref.1 (len:46824501) : chr21_hapA\n",
      "ref.2 (len:46818423) : chr21_hapB\n",
      "\n",
      ":::: Simulation stats (ref.1) ::::\n",
      "\n",
      "read num. : 10473\n",
      "depth : 4.500005\n",
      "read length mean (SD) : 20119.401413 (6927.139265)\n",
      "read length min : 3693\n",
      "read length max : 54348\n",
      "read accuracy mean (SD) : 0.960128 (0.042890)\n",
      "substitution rate. : 0.002401\n",
      "insertion rate. : 0.022033\n",
      "deletion rate. : 0.015606\n",
      "\n",
      "\n",
      "Merging haplotypes\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.801*0.97] collected minimizers\n",
      "[M::mm_idx_gen::0.853*1.29] sorted minimizers\n",
      "[M::main::0.853*1.29] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.893*1.27] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.921*1.26] distinct minimizers: 3249742 (93.82% are singletons); average occurrences: 1.236; average spacing: 11.627; total length: 46709983\n",
      "[M::worker_pipeline::25.876*5.81] mapped 20903 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax map-hifi ./clones/chr21.fa ./clones//dependency_0//sim_lr.pbsim.fastq\n",
      "[M::main] Real time: 25.891 sec; CPU: 150.366 sec; Peak RSS: 1.879 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0///A_sim_lr.pbsim.bam > ./clones//dependency_0///A_sim_lr.pbsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0///A_sim_lr.pbsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t18572\t40058265\t85.7595\t7.62585\t30.2\t56.9\n",
      "\n",
      "Simulating reads for clone B\n",
      "Clone purity 20 Clone coverage 1.5 30 20 2\n",
      "STDERR: :::: Simulation parameters :::\n",
      "\n",
      "strategy : wgs\n",
      "method : qshmm\n",
      "qshmm : /Users/ebattist/anaconda3/envs/insilicosv-demo-env/data/QSHMM-RSII.model\n",
      "genome : ./clones//dependency_0//dependency_1///sim.fa\n",
      "prefix : ./clones//dependency_0//dependency_1//sim_lr.pbsim\n",
      "id-prefix : S\n",
      "depth : 1.500000\n",
      "length-mean : 20000.000000\n",
      "length-sd : 7000.000000\n",
      "length-min : 100\n",
      "length-max : 1000000\n",
      "difference-ratio : 6:55:39\n",
      "seed : 1755008620\n",
      "accuracy-mean : 0.990000\n",
      "pass_num : 1\n",
      "hp-del-bias : 1.000000\n",
      "\n",
      ":::: Reference stats ::::\n",
      "\n",
      "file name : ./clones//dependency_0//dependency_1///sim.fa\n",
      "\n",
      "ref.1 (len:46824613) : chr21_hapA_hapA\n",
      "ref.2 (len:46818308) : chr21_hapB_hapA\n",
      "ref.3 (len:46824817) : chr21_hapA_hapB\n",
      "ref.4 (len:46818182) : chr21_hapB_hapB\n",
      "\n",
      ":::: Simulation stats (ref.1) ::::\n",
      "\n",
      "read num. : 3470\n",
      "depth : 1.500002\n",
      "read length mean (SD) : 20241.216427 (7104.582449)\n",
      "read length min : 4198\n",
      "read length max : 48605\n",
      "read accuracy mean (SD) : 0.960428 (0.042472)\n",
      "substitution rate. : 0.002378\n",
      "insertion rate. : 0.021800\n",
      "deletion rate. : 0.015388\n",
      "\n",
      "\n",
      "Merging haplotypes\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.789*0.98] collected minimizers\n",
      "[M::mm_idx_gen::0.845*1.29] sorted minimizers\n",
      "[M::main::0.846*1.29] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.884*1.28] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.913*1.27] distinct minimizers: 3249742 (93.82% are singletons); average occurrences: 1.236; average spacing: 11.627; total length: 46709983\n",
      "[M::worker_pipeline::9.286*5.53] mapped 6950 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax map-hifi ./clones/chr21.fa ./clones//dependency_0//dependency_1//sim_lr.pbsim.fastq\n",
      "[M::main] Real time: 9.299 sec; CPU: 51.392 sec; Peak RSS: 1.944 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1///B_sim_lr.pbsim.bam > ./clones//dependency_0//dependency_1///B_sim_lr.pbsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1///B_sim_lr.pbsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t6157\t37944731\t81.2347\t2.5316\t30.5\t56.9\n",
      "\n",
      "Simulating reads for clone C\n",
      "Clone purity 20 Clone coverage 1.0 30 20 3\n",
      "STDERR: :::: Simulation parameters :::\n",
      "\n",
      "strategy : wgs\n",
      "method : qshmm\n",
      "qshmm : /Users/ebattist/anaconda3/envs/insilicosv-demo-env/data/QSHMM-RSII.model\n",
      "genome : ./clones//dependency_0//dependency_1//dependency_2///sim.fa\n",
      "prefix : ./clones//dependency_0//dependency_1//dependency_2//sim_lr.pbsim\n",
      "id-prefix : S\n",
      "depth : 1.000000\n",
      "length-mean : 20000.000000\n",
      "length-sd : 7000.000000\n",
      "length-min : 100\n",
      "length-max : 1000000\n",
      "difference-ratio : 6:55:39\n",
      "seed : 1755008654\n",
      "accuracy-mean : 0.990000\n",
      "pass_num : 1\n",
      "hp-del-bias : 1.000000\n",
      "\n",
      ":::: Reference stats ::::\n",
      "\n",
      "file name : ./clones//dependency_0//dependency_1//dependency_2///sim.fa\n",
      "\n",
      "ref.1 (len:46807604) : chr21_hapA_hapA_hapA\n",
      "ref.2 (len:46816024) : chr21_hapB_hapA_hapA\n",
      "ref.3 (len:46831148) : chr21_hapA_hapB_hapA\n",
      "ref.4 (len:46809734) : chr21_hapB_hapB_hapA\n",
      "ref.5 (len:46828375) : chr21_hapA_hapA_hapB\n",
      "ref.6 (len:46822204) : chr21_hapB_hapA_hapB\n",
      "ref.7 (len:46822836) : chr21_hapA_hapB_hapB\n",
      "ref.8 (len:46809734) : chr21_hapB_hapB_hapB\n",
      "\n",
      ":::: Simulation stats (ref.1) ::::\n",
      "\n",
      "read num. : 2318\n",
      "depth : 1.000002\n",
      "read length mean (SD) : 20193.137619 (7084.436125)\n",
      "read length min : 4554\n",
      "read length max : 50184\n",
      "read accuracy mean (SD) : 0.960105 (0.042667)\n",
      "substitution rate. : 0.002401\n",
      "insertion rate. : 0.022080\n",
      "deletion rate. : 0.015609\n",
      "\n",
      "\n",
      "Merging haplotypes\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.721*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.769*1.35] sorted minimizers\n",
      "[M::main::0.769*1.35] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.805*1.33] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.832*1.32] distinct minimizers: 3249742 (93.82% are singletons); average occurrences: 1.236; average spacing: 11.627; total length: 46709983\n",
      "[M::worker_pipeline::6.570*4.88] mapped 4658 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax map-hifi ./clones/chr21.fa ./clones//dependency_0//dependency_1//dependency_2//sim_lr.pbsim.fastq\n",
      "[M::main] Real time: 6.585 sec; CPU: 32.046 sec; Peak RSS: 1.690 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1//dependency_2///C_sim_lr.pbsim.bam > ./clones//dependency_0//dependency_1//dependency_2///C_sim_lr.pbsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1//dependency_2///C_sim_lr.pbsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t4127\t34207060\t73.2329\t1.68268\t30.9\t57.1\n",
      "\n",
      "Simulating reads for clone D\n",
      "Clone purity 20 Clone coverage 1.0 30 20 3\n",
      "STDERR: :::: Simulation parameters :::\n",
      "\n",
      "strategy : wgs\n",
      "method : qshmm\n",
      "qshmm : /Users/ebattist/anaconda3/envs/insilicosv-demo-env/data/QSHMM-RSII.model\n",
      "genome : ./clones//dependency_0//dependency_1//dependency_3///sim.fa\n",
      "prefix : ./clones//dependency_0//dependency_1//dependency_3//sim_lr.pbsim\n",
      "id-prefix : S\n",
      "depth : 1.000000\n",
      "length-mean : 20000.000000\n",
      "length-sd : 7000.000000\n",
      "length-min : 100\n",
      "length-max : 1000000\n",
      "difference-ratio : 6:55:39\n",
      "seed : 1755008676\n",
      "accuracy-mean : 0.990000\n",
      "pass_num : 1\n",
      "hp-del-bias : 1.000000\n",
      "\n",
      ":::: Reference stats ::::\n",
      "\n",
      "file name : ./clones//dependency_0//dependency_1//dependency_3///sim.fa\n",
      "\n",
      "ref.1 (len:46830189) : chr21_hapA_hapA_hapA\n",
      "ref.2 (len:46818308) : chr21_hapB_hapA_hapA\n",
      "ref.3 (len:46824817) : chr21_hapA_hapB_hapA\n",
      "ref.4 (len:46824419) : chr21_hapB_hapB_hapA\n",
      "ref.5 (len:46824613) : chr21_hapA_hapA_hapB\n",
      "ref.6 (len:46826082) : chr21_hapB_hapA_hapB\n",
      "ref.7 (len:46824817) : chr21_hapA_hapB_hapB\n",
      "ref.8 (len:46833820) : chr21_hapB_hapB_hapB\n",
      "\n",
      ":::: Simulation stats (ref.1) ::::\n",
      "\n",
      "read num. : 2327\n",
      "depth : 1.000000\n",
      "read length mean (SD) : 20124.705200 (6864.335586)\n",
      "read length min : 4800\n",
      "read length max : 49605\n",
      "read accuracy mean (SD) : 0.960682 (0.042067)\n",
      "substitution rate. : 0.002372\n",
      "insertion rate. : 0.021699\n",
      "deletion rate. : 0.015345\n",
      "\n",
      "\n",
      "Merging haplotypes\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.765*0.99] collected minimizers\n",
      "[M::mm_idx_gen::0.815*1.32] sorted minimizers\n",
      "[M::main::0.816*1.32] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.855*1.30] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.886*1.29] distinct minimizers: 3249742 (93.82% are singletons); average occurrences: 1.236; average spacing: 11.627; total length: 46709983\n",
      "[M::worker_pipeline::6.427*5.21] mapped 4658 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax map-hifi ./clones/chr21.fa ./clones//dependency_0//dependency_1//dependency_3//sim_lr.pbsim.fastq\n",
      "[M::main] Real time: 6.441 sec; CPU: 33.487 sec; Peak RSS: 1.705 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_1//dependency_3///D_sim_lr.pbsim.bam > ./clones//dependency_0//dependency_1//dependency_3///D_sim_lr.pbsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_1//dependency_3///D_sim_lr.pbsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t4105\t34393817\t73.6327\t1.68255\t30.9\t56.9\n",
      "\n",
      "Simulating reads for clone E\n",
      "Clone purity 10 Clone coverage 0.75 30 10 2\n",
      "STDERR: :::: Simulation parameters :::\n",
      "\n",
      "strategy : wgs\n",
      "method : qshmm\n",
      "qshmm : /Users/ebattist/anaconda3/envs/insilicosv-demo-env/data/QSHMM-RSII.model\n",
      "genome : ./clones//dependency_0//dependency_4///sim.fa\n",
      "prefix : ./clones//dependency_0//dependency_4//sim_lr.pbsim\n",
      "id-prefix : S\n",
      "depth : 0.750000\n",
      "length-mean : 20000.000000\n",
      "length-sd : 7000.000000\n",
      "length-min : 100\n",
      "length-max : 1000000\n",
      "difference-ratio : 6:55:39\n",
      "seed : 1755008699\n",
      "accuracy-mean : 0.990000\n",
      "pass_num : 1\n",
      "hp-del-bias : 1.000000\n",
      "\n",
      ":::: Reference stats ::::\n",
      "\n",
      "file name : ./clones//dependency_0//dependency_4///sim.fa\n",
      "\n",
      "ref.1 (len:46824501) : chr21_hapA_hapA\n",
      "ref.2 (len:46818423) : chr21_hapB_hapA\n",
      "ref.3 (len:46824501) : chr21_hapA_hapB\n",
      "ref.4 (len:46818423) : chr21_hapB_hapB\n",
      "\n",
      ":::: Simulation stats (ref.1) ::::\n",
      "\n",
      "read num. : 1739\n",
      "depth : 0.750000\n",
      "read length mean (SD) : 20194.580219 (7053.186132)\n",
      "read length min : 4450\n",
      "read length max : 50156\n",
      "read accuracy mean (SD) : 0.959520 (0.042703)\n",
      "substitution rate. : 0.002418\n",
      "insertion rate. : 0.022143\n",
      "deletion rate. : 0.015675\n",
      "\n",
      "\n",
      "Merging haplotypes\n",
      "Aligning reads\n",
      "STDERR: [M::mm_idx_gen::0.714*1.02] collected minimizers\n",
      "[M::mm_idx_gen::0.761*1.38] sorted minimizers\n",
      "[M::main::0.761*1.38] loaded/built the index for 1 target sequence(s)\n",
      "[M::mm_mapopt_update::0.800*1.36] mid_occ = 104\n",
      "[M::mm_idx_stat] kmer size: 19; skip: 19; is_hpc: 0; #seq: 1\n",
      "[M::mm_idx_stat::0.826*1.35] distinct minimizers: 3249742 (93.82% are singletons); average occurrences: 1.236; average spacing: 11.627; total length: 46709983\n",
      "[M::worker_pipeline::4.550*5.64] mapped 3496 sequences\n",
      "[M::main] Version: 2.30-r1287\n",
      "[M::main] CMD: minimap2 -t 10 -ax map-hifi ./clones/chr21.fa ./clones//dependency_0//dependency_4//sim_lr.pbsim.fastq\n",
      "[M::main] Real time: 4.559 sec; CPU: 25.671 sec; Peak RSS: 1.567 GB\n",
      "[bam_sort_core] merging from 0 files and 10 in-memory blocks...\n",
      "\n",
      "samtools coverage ./clones//dependency_0//dependency_4///E_sim_lr.pbsim.bam > ./clones//dependency_0//dependency_4///E_sim_lr.pbsim.bam.coverage\n",
      "PROCESSED: Clone ./clones//dependency_0//dependency_4///E_sim_lr.pbsim.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t3106\t31135122\t66.6562\t1.26056\t30.2\t57.1\n",
      "\n"
     ]
    }
   ],
   "execution_count": 85
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "ExecuteTime": {
     "end_time": "2025-08-12T14:28:22.545482Z",
     "start_time": "2025-08-12T14:28:09.917322Z"
    }
   },
   "source": [
    "output_name_lr = root_path + 'cancer_genome_lr.bam'\n",
    "merge_clones(output_name_lr, list_lr_bams)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merging clones...\n",
      "Indexing reads...\n",
      "Computing coverage...\n",
      "PROCESSED: Whole cancer genome ./clones/cancer_genome_lr.bam coverage #rname\tstartpos\tendpos\tnumreads\tcovbases\tcoverage\tmeandepth\tmeanbaseq\tmeanmapq\n",
      "chr21\t1\t46709983\t36067\t40071294\t85.7874\t14.7832\t30.4\t57\n",
      "\n"
     ]
    }
   ],
   "execution_count": 89
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": [
    "## SV Visualization\n",
    "Below we generate ```samplot``` illustrations for each SV simulated with ```insilicoSV```."
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    },
    "scrolled": false,
    "ExecuteTime": {
     "end_time": "2025-08-12T14:36:52.542280Z",
     "start_time": "2025-08-12T14:36:45.662736Z"
    }
   },
   "source": [
    "samplot_path = root_path + 'samplot/'\n",
    "os.makedirs(samplot_path, exist_ok=True)\n",
    "\n",
    "os.environ[\"MPLBACKEND\"] = \"Agg\"\n",
    "\n",
    "for vcf_path in vcf_list:\n",
    "    vcf = VariantFile(vcf_path)\n",
    "    rec2breakends = defaultdict(set)\n",
    "    for vcf_rec in vcf.fetch():\n",
    "        vcf_info = dict(vcf_rec.info)        \n",
    "        sv_title = \"%s_%s_%s\" % (vcf_info.get('SVTYPE', 'SNP'), \n",
    "                                 vcf_info.get('SVID', vcf_rec.id), \n",
    "                                 vcf_info.get('GRAMMAR', '').replace(\"->\", \"-to-\"))\n",
    "        rec2breakends[sv_title].add(vcf_rec.start)\n",
    "        rec2breakends[sv_title].add(vcf_rec.stop)\n",
    "        if 'TARGET' in vcf_info:\n",
    "            rec2breakends[sv_title].add(vcf_info['TARGET'])\n",
    "    vcf.close()\n",
    "               \n",
    "    for sv_title, sv_breakends in rec2breakends.items():\n",
    "        sv_breakends = sorted(sv_breakends)\n",
    "        output_file = \"%s/%s.png\" % (samplot_path, sv_title)\n",
    "        start = min(sv_breakends)\n",
    "        end = max(sv_breakends)\n",
    "        wlen = end - start + 1000\n",
    "        command = [\"samplot plot -n Illumina HiFi -b %s %s -s %s -e %s -c chr21 -t %s -w %d\" \\\n",
    "              \" --include_mqual 0 --separate_mqual 1 -o %s\" % (output_name_sr, output_name_lr, \n",
    "                                                               start, end, sv_title, wlen, output_file)]\n",
    "        run_bash_process(command)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 6664646 -e 6671835 -c chr21 -t DUP_INV_sv17_A-to-aa -w 8189 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_INV_sv17_A-to-aa.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 8600331 -e 8609963 -c chr21 -t INV_sv13_A-to-a -w 10632 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INV_sv13_A-to-a.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 13373438 -e 13388315 -c chr21 -t CUSTOM_sv21_AB-to-AbbA -w 15877 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//CUSTOM_sv21_AB-to-AbbA.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 13757028 -e 13763976 -c chr21 -t DUP_sv6_A-to-A+A -w 7948 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv6_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 13780748 -e 13789090 -c chr21 -t DUP_INV_sv15_A-to-aa -w 9342 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_INV_sv15_A-to-aa.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 14236945 -e 14246590 -c chr21 -t DUP_INV_sv18_A-to-aa -w 10645 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_INV_sv18_A-to-aa.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 18279330 -e 18289275 -c chr21 -t DEL_sv1_A-to- -w 10945 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv1_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 19061436 -e 19070195 -c chr21 -t DEL_sv2_A-to- -w 9759 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv2_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 23474788 -e 23481674 -c chr21 -t INV_sv11_A-to-a -w 7886 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INV_sv11_A-to-a.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24282571 -e 24298353 -c chr21 -t CUSTOM_sv22_AB-to-AbbA -w 16782 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//CUSTOM_sv22_AB-to-AbbA.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24599497 -e 24617027 -c chr21 -t CUSTOM_sv20_AB-to-AbbA -w 18530 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//CUSTOM_sv20_AB-to-AbbA.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27433938 -e 27443403 -c chr21 -t DUP_INV_sv19_A-to-aa -w 10465 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_INV_sv19_A-to-aa.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27552590 -e 27561327 -c chr21 -t INV_sv14_A-to-a -w 9737 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INV_sv14_A-to-a.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27739464 -e 27745544 -c chr21 -t INV_sv12_A-to-a -w 7080 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INV_sv12_A-to-a.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 28863467 -e 28881867 -c chr21 -t CUSTOM_sv24_AB-to-AbbA -w 19400 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//CUSTOM_sv24_AB-to-AbbA.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29097628 -e 29105847 -c chr21 -t DEL_sv3_A-to- -w 9219 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv3_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29170860 -e 29188549 -c chr21 -t CUSTOM_sv23_AB-to-AbbA -w 18689 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//CUSTOM_sv23_AB-to-AbbA.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29314353 -e 29323952 -c chr21 -t DUP_sv8_A-to-A+A -w 10599 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv8_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 31169722 -e 31176744 -c chr21 -t DUP_INV_sv16_A-to-aa -w 8022 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_INV_sv16_A-to-aa.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33788903 -e 33798402 -c chr21 -t DUP_sv5_A-to-A+A -w 10499 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv5_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 35705259 -e 35714337 -c chr21 -t DUP_sv9_A-to-A+A -w 10078 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv9_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 37072348 -e 37081098 -c chr21 -t DEL_sv0_A-to- -w 9750 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv0_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 37514311 -e 37523811 -c chr21 -t DEL_sv4_A-to- -w 10500 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv4_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 38673242 -e 38679818 -c chr21 -t INV_sv10_A-to-a -w 7576 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INV_sv10_A-to-a.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 43043514 -e 43050040 -c chr21 -t DUP_sv7_A-to-AA+ -w 7526 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv7_A-to-AA+.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 5891597 -e 5891637 -c chr21 -t INDEL_sv13_A-to- -w 1040 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv13_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 9578598 -e 9578599 -c chr21 -t INDEL_sv20_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv20_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 9954332 -e 9954369 -c chr21 -t INDEL_sv24_A-to- -w 1037 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv24_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 12892643 -e 12892644 -c chr21 -t INDEL_sv45_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv45_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 15807270 -e 15807300 -c chr21 -t INDEL_sv30_A-to- -w 1030 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv30_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 17003225 -e 17003226 -c chr21 -t INDEL_sv23_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv23_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 20335059 -e 20335092 -c chr21 -t INDEL_sv40_A-to- -w 1033 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv40_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21091684 -e 21091685 -c chr21 -t INDEL_sv35_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv35_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21926813 -e 21926860 -c chr21 -t INDEL_sv36_A-to- -w 1047 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv36_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24826763 -e 24826764 -c chr21 -t INDEL_sv34_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv34_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24955421 -e 24955422 -c chr21 -t INDEL_sv22_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv22_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27037557 -e 27037602 -c chr21 -t INDEL_sv5_A-to- -w 1045 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv5_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27799941 -e 27799942 -c chr21 -t INDEL_sv42_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv42_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27814073 -e 27814074 -c chr21 -t INDEL_sv38_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv38_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 28360723 -e 28360724 -c chr21 -t INDEL_sv17_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv17_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 30609968 -e 30609969 -c chr21 -t INDEL_sv28_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv28_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 31806533 -e 31806575 -c chr21 -t INDEL_sv8_A-to- -w 1042 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv8_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 32091825 -e 32091826 -c chr21 -t INDEL_sv0_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv0_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 36673063 -e 36673110 -c chr21 -t INDEL_sv12_A-to- -w 1047 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv12_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 36963620 -e 36963621 -c chr21 -t INDEL_sv4_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv4_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 37862020 -e 37862021 -c chr21 -t INDEL_sv27_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv27_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 38951610 -e 38951611 -c chr21 -t INDEL_sv31_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv31_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 39978673 -e 39978705 -c chr21 -t INDEL_sv16_A-to- -w 1032 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv16_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 42266353 -e 42266389 -c chr21 -t INDEL_sv21_A-to- -w 1036 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv21_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 43046606 -e 43046607 -c chr21 -t INDEL_sv10_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv10_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 44251583 -e 44251622 -c chr21 -t INDEL_sv32_A-to- -w 1039 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv32_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 45349927 -e 45349928 -c chr21 -t INDEL_sv49_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv49_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 45490352 -e 45490392 -c chr21 -t INDEL_sv29_A-to- -w 1040 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv29_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 45697647 -e 45697648 -c chr21 -t INDEL_sv39_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv39_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 45872717 -e 45872764 -c chr21 -t INDEL_sv15_A-to- -w 1047 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv15_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 17482033 -e 17482034 -c chr21 -t INDEL_sv37_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv37_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 19589817 -e 19589818 -c chr21 -t INDEL_sv47_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv47_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21036968 -e 21037016 -c chr21 -t INDEL_sv2_A-to- -w 1048 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv2_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22640176 -e 22640177 -c chr21 -t INDEL_sv41_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv41_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24293476 -e 24293516 -c chr21 -t INDEL_sv14_A-to- -w 1040 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv14_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 25848181 -e 25848217 -c chr21 -t INDEL_sv44_A-to- -w 1036 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv44_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 28104378 -e 28104379 -c chr21 -t INDEL_sv48_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv48_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29647741 -e 29647776 -c chr21 -t INDEL_sv46_A-to- -w 1035 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv46_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29740894 -e 29740936 -c chr21 -t INDEL_sv19_A-to- -w 1042 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv19_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29926829 -e 29926830 -c chr21 -t INDEL_sv26_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv26_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 32227179 -e 32227213 -c chr21 -t INDEL_sv11_A-to- -w 1034 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv11_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33722781 -e 33722782 -c chr21 -t INDEL_sv25_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv25_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 36816059 -e 36816060 -c chr21 -t INDEL_sv33_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv33_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 39988932 -e 39988978 -c chr21 -t INDEL_sv18_A-to- -w 1046 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv18_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 40594324 -e 40594364 -c chr21 -t INDEL_sv3_A-to- -w 1040 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv3_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 41487974 -e 41488010 -c chr21 -t INDEL_sv6_A-to- -w 1036 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv6_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 41591705 -e 41591746 -c chr21 -t INDEL_sv7_A-to- -w 1041 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv7_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 41641719 -e 41641762 -c chr21 -t INDEL_sv1_A-to- -w 1043 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv1_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 42666438 -e 42666482 -c chr21 -t INDEL_sv9_A-to- -w 1044 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv9_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 42866460 -e 42866461 -c chr21 -t INDEL_sv43_-to-A -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//INDEL_sv43_-to-A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 7501919 -e 7501920 -c chr21 -t SNP_snp30_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp30_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 10370737 -e 10377286 -c chr21 -t DUP_sv5_A-to-A+A -w 7549 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv5_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 10694672 -e 10694673 -c chr21 -t SNP_snp14_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp14_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 11410699 -e 11410700 -c chr21 -t SNP_snp32_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp32_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 14172100 -e 14172101 -c chr21 -t SNP_snp27_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp27_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 15678470 -e 15678471 -c chr21 -t SNP_snp15_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp15_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 17786654 -e 17792484 -c chr21 -t DUP_sv8_A-to-A+A -w 6830 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv8_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22489715 -e 22489716 -c chr21 -t SNP_snp34_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp34_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22860625 -e 22860626 -c chr21 -t SNP_snp42_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp42_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 23268782 -e 23277177 -c chr21 -t DEL_sv4_A-to- -w 9395 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv4_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 23594067 -e 23594068 -c chr21 -t SNP_snp18_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp18_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 30112027 -e 30112028 -c chr21 -t SNP_snp40_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp40_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 31631972 -e 31631973 -c chr21 -t SNP_snp29_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp29_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 32725760 -e 32725761 -c chr21 -t SNP_snp31_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp31_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 34081171 -e 34089787 -c chr21 -t DEL_sv1_A-to- -w 9616 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv1_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 39720652 -e 39720653 -c chr21 -t SNP_snp21_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp21_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 42591558 -e 42591559 -c chr21 -t SNP_snp35_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp35_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 45277183 -e 45277184 -c chr21 -t SNP_snp52_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp52_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 8823559 -e 8823560 -c chr21 -t SNP_snp17_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp17_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 9844931 -e 9844932 -c chr21 -t SNP_snp58_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp58_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 13550620 -e 13550621 -c chr21 -t SNP_snp25_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp25_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 14470951 -e 14470952 -c chr21 -t SNP_snp38_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp38_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 15754248 -e 15754249 -c chr21 -t SNP_snp59_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp59_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 18560240 -e 18560241 -c chr21 -t SNP_snp48_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp48_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22919413 -e 22925745 -c chr21 -t DUP_sv7_A-to-AA+ -w 7332 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv7_A-to-AA+.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 30853650 -e 30853651 -c chr21 -t SNP_snp37_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp37_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33010861 -e 33010862 -c chr21 -t SNP_snp24_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp24_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33022507 -e 33022508 -c chr21 -t SNP_snp28_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp28_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 35975814 -e 35975815 -c chr21 -t SNP_snp51_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp51_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 38360881 -e 38360882 -c chr21 -t SNP_snp45_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp45_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 43067005 -e 43075318 -c chr21 -t DEL_sv0_A-to- -w 9313 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv0_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 44273279 -e 44273280 -c chr21 -t SNP_snp11_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp11_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 44539354 -e 44539355 -c chr21 -t SNP_snp49_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp49_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 20122103 -e 20122104 -c chr21 -t SNP_snp57_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp57_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21026636 -e 21026637 -c chr21 -t SNP_snp36_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp36_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21886433 -e 21886434 -c chr21 -t SNP_snp13_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp13_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22536552 -e 22536553 -c chr21 -t SNP_snp43_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp43_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 23767350 -e 23767351 -c chr21 -t SNP_snp54_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp54_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 24570316 -e 24576497 -c chr21 -t DUP_sv6_A-to-AA+ -w 7181 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv6_A-to-AA+.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 28359661 -e 28359662 -c chr21 -t SNP_snp41_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp41_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 29674283 -e 29674284 -c chr21 -t SNP_snp47_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp47_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 30858395 -e 30858396 -c chr21 -t SNP_snp20_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp20_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 31716180 -e 31716181 -c chr21 -t SNP_snp12_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp12_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33822253 -e 33822254 -c chr21 -t SNP_snp53_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp53_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 38704545 -e 38712181 -c chr21 -t DEL_sv3_A-to- -w 8636 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv3_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 41977309 -e 41977310 -c chr21 -t SNP_snp50_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp50_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 42145699 -e 42145700 -c chr21 -t SNP_snp56_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp56_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 44834460 -e 44839812 -c chr21 -t DUP_sv9_A-to-A+A -w 6352 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DUP_sv9_A-to-A+A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 5151945 -e 5151946 -c chr21 -t SNP_snp23_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp23_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 5365430 -e 5365431 -c chr21 -t SNP_snp44_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp44_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 6742799 -e 6742800 -c chr21 -t SNP_snp19_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp19_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 12381450 -e 12389899 -c chr21 -t DEL_sv2_A-to- -w 9449 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//DEL_sv2_A-to-.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 13271164 -e 13271165 -c chr21 -t SNP_snp39_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp39_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 14742793 -e 14742794 -c chr21 -t SNP_snp22_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp22_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 21947413 -e 21947414 -c chr21 -t SNP_snp26_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp26_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 26392446 -e 26392447 -c chr21 -t SNP_snp10_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp10_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27765042 -e 27765043 -c chr21 -t SNP_snp16_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp16_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 32688360 -e 32688361 -c chr21 -t SNP_snp46_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp46_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 39753417 -e 39753418 -c chr21 -t SNP_snp33_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp33_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 40268998 -e 40268999 -c chr21 -t SNP_snp55_ -w 1001 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//SNP_snp55_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 27733778 -e 27745342 -c chr21 -t dDUP_sv2__A-to-A_A -w 12564 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//dDUP_sv2__A-to-A_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 17022285 -e 17036231 -c chr21 -t dDUP_sv0__A-to-A_A -w 14946 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//dDUP_sv0__A-to-A_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 22698329 -e 22714009 -c chr21 -t dDUP_sv3__A-to-A_A -w 16680 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//dDUP_sv3__A-to-A_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 26753823 -e 26769929 -c chr21 -t dDUP_sv1__A-to-A_A -w 17106 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//dDUP_sv1__A-to-A_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 43545804 -e 43562949 -c chr21 -t dDUP_sv4_A_-to-A_A -w 18145 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//dDUP_sv4_A_-to-A_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 25149900 -e 25169569 -c chr21 -t nrTRA_sv0_A_-to-_A -w 20669 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//nrTRA_sv0_A_-to-_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 26407909 -e 26422659 -c chr21 -t nrTRA_sv3__A-to-A_ -w 15750 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//nrTRA_sv3__A-to-A_.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 32137574 -e 32151679 -c chr21 -t nrTRA_sv1_A_-to-_A -w 15105 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//nrTRA_sv1_A_-to-_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 11624895 -e 11640846 -c chr21 -t nrTRA_sv4_A_-to-_A -w 16951 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//nrTRA_sv4_A_-to-_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n",
      "['samplot plot -n Illumina HiFi -b ./clones/cancer_genome_sr.bam ./clones/cancer_genome_lr.bam -s 33889163 -e 33902005 -c chr21 -t nrTRA_sv2_A_-to-_A -w 13842 --include_mqual 0 --separate_mqual 1 -o ./clones/samplot//nrTRA_sv2_A_-to-_A.png']\n",
      "STDERR: /bin/sh: samplot: command not found\n",
      "\n"
     ]
    }
   ],
   "execution_count": 92
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": [
    "%%sh\n",
    "ls -l clones/samplot/"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "pycharm": {
     "name": "#%% md\n"
    }
   },
   "source": "We visualize some SVs below."
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": "Image(filename=samplot_path + '/DEL_sv0_A-to-.png')",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": "Image(filename=samplot_path + 'DUP_sv5_A-to-AA+.png')",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": "Image(filename=samplot_path + 'INV_sv10_A-to-a.png')",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "pycharm": {
     "name": "#%%\n"
    }
   },
   "source": "Image(filename=samplot_path + 'Custom_sv15_A-to-aa.png')",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": "Image(filename=samplot_path + 'Custom_sv20_AB-to-AbbA.png')",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (insislicoSV)",
   "language": "python",
   "name": "insilicosv"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
